{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyP1LVeXHSP5zE3xtFwthbgb",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Deusxy/Ribosome-Binding-Site-Calculator-v1.0/blob/master/RBSCal_Project.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Project \n",
        "Side project get the RBS calculater runing as well as the disign. \n",
        "https://github.com/Deusxy/Ribosome-Binding-Site-Calculator-v1.0 "
      ],
      "metadata": {
        "id": "SznDVdMWlvyK"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "woWX7r0uRNLL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title code of ViennaRNA\n",
        "#This file is part of the Ribosome Binding Site Calculator.\n",
        "#The Ribosome Binding Site Calculator is free software: you can redistribute it and/or modify\n",
        "\n",
        "import os.path\n",
        "import os, popen2, time\n",
        "\n",
        "current_dir = os.path.dirname(os.path.abspath(__file__)) + \"/tmp\"\n",
        "if not os.path.exists(current_dir): os.mkdir(current_dir)\n",
        "\n",
        "debug=0\n",
        "\n",
        "#Class that encapsulates all of the functions from NuPACK 2.0\n",
        "class ViennaRNA(dict):\n",
        "\n",
        "    debug_mode = 0\n",
        "    RT = 0.61597 #gas constant times 310 Kelvin (in units of kcal/mol)\n",
        "\n",
        "    def __init__(self,Sequence_List,material = \"rna37\"):\n",
        "\n",
        "        self.ran = 0\n",
        "\n",
        "        import re\n",
        "        import random\n",
        "        import string\n",
        "\n",
        "        exp = re.compile('[ATGCU]',re.IGNORECASE)\n",
        "\n",
        "        for seq in Sequence_List:\n",
        "            if exp.match(seq) == None:\n",
        "                error_string = \"Invalid letters found in inputted sequences. Only ATGCU allowed. \\n Sequence is \\\"\" + str(seq) + \"\\\".\"\n",
        "                raise ValueError(error_string)\n",
        "\n",
        "        self[\"sequences\"] = Sequence_List\n",
        "        self[\"material\"] = material\n",
        "\n",
        "        random.seed(time.time())\n",
        "        long_id = \"\".join([random.choice(string.letters + string.digits) for x in range(10)])\n",
        "        self.prefix = current_dir + \"/temp_\" + long_id\n",
        "\n",
        "    def mfe(self, strands,Temp = 37.0, dangles = \"all\",outputPS = False):\n",
        "\n",
        "        self[\"mfe_composition\"] = strands\n",
        "\n",
        "        if Temp <= 0: raise ValueError(\"The specified temperature must be greater than zero.\")\n",
        "\n",
        "        seq_string = \"&\".join(self[\"sequences\"])\n",
        "        input_string = seq_string + \"\\n\"\n",
        "\n",
        "        handle = open(self.prefix,\"w\")\n",
        "        handle.write(input_string)\n",
        "        handle.close()\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if dangles is \"none\":\n",
        "            dangles = \" -d0 \"\n",
        "        elif dangles is \"some\":\n",
        "            dangles = \" -d1 \"\n",
        "        elif dangles is \"all\":\n",
        "            dangles = \" -d2 \"\n",
        "\n",
        "        if outputPS:\n",
        "            outputPS_str = \" \"\n",
        "        else:\n",
        "            outputPS_str = \" -noPS \"\n",
        "\n",
        "        #Call ViennaRNA C programs\n",
        "        cmd = \"RNAcofold\"\n",
        "        args = outputPS_str + dangles + \" < \" + self.prefix\n",
        "\n",
        "        output = popen2.Popen3(cmd + args)\n",
        "        #output.tochild.write(input_string)\n",
        "\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        #Skip the unnecessary output lines\n",
        "        line = output.fromchild.readline()\n",
        "\n",
        "        line = output.fromchild.readline()\n",
        "        words = line.split(\" \")\n",
        "        bracket_string = words[0]\n",
        "        (strands,bp_x, bp_y) = self.convert_bracket_to_numbered_pairs(bracket_string)\n",
        "\n",
        "        energy = float(words[len(words)-1].replace(\")\",\"\").replace(\"(\",\"\").replace(\"\\n\",\"\"))\n",
        "\n",
        "        self._cleanup()\n",
        "        self[\"program\"] = \"mfe\"\n",
        "        self[\"mfe_basepairing_x\"] = [bp_x]\n",
        "        self[\"mfe_basepairing_y\"] = [bp_y]\n",
        "        self[\"mfe_energy\"] = [energy]\n",
        "        self[\"totalnt\"]=strands\n",
        "\n",
        "\n",
        "        #print \"Minimum free energy secondary structure has been calculated.\"\n",
        "\n",
        "    def subopt(self, strands,energy_gap,Temp = 37.0, dangles = \"all\", outputPS = False):\n",
        "\n",
        "        self[\"subopt_composition\"] = strands\n",
        "\n",
        "        if Temp <= 0: raise ValueError(\"The specified temperature must be greater than zero.\")\n",
        "\n",
        "        seq_string = \"&\".join(self[\"sequences\"])\n",
        "        input_string = seq_string + \"\\n\"\n",
        "\n",
        "        handle = open(self.prefix,\"w\")\n",
        "        handle.write(input_string)\n",
        "        handle.close()\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if dangles is \"none\":\n",
        "            dangles = \" -d0 \"\n",
        "        elif dangles is \"some\":\n",
        "            dangles = \" -d1 \"\n",
        "        elif dangles is \"all\":\n",
        "            dangles = \" -d2 \"\n",
        "\n",
        "        if outputPS:\n",
        "            outputPS_str = \" \"\n",
        "        else:\n",
        "            outputPS_str = \" -noPS \"\n",
        "\n",
        "        #Call ViennaRNA C programs\n",
        "        cmd = \"RNAsubopt\"\n",
        "        args = \" -e \" + str(energy_gap) + outputPS_str + dangles + \" < \" + self.prefix\n",
        "\n",
        "        output = popen2.Popen3(cmd + args)\n",
        "\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        #print output.fromchild.read()\n",
        "        if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        #Skip unnecessary line\n",
        "        line = output.fromchild.readline()\n",
        "\n",
        "        self[\"subopt_basepairing_x\"] = []\n",
        "        self[\"subopt_basepairing_y\"] = []\n",
        "        self[\"subopt_energy\"] = []\n",
        "        self[\"totalnt\"]=[]\n",
        "        counter=0\n",
        "\n",
        "        while len(line)>0:\n",
        "            line = output.fromchild.readline()\n",
        "            if len(line) > 0:\n",
        "                counter+=1\n",
        "                words = line.split(\" \")\n",
        "                bracket_string = words[0]\n",
        "                energy = float(words[len(words)-1].replace(\"\\n\",\"\"))\n",
        "\n",
        "                (strands,bp_x, bp_y) = self.convert_bracket_to_numbered_pairs(bracket_string)\n",
        "\n",
        "                self[\"subopt_energy\"].append(energy)\n",
        "                self[\"subopt_basepairing_x\"].append(bp_x)\n",
        "                self[\"subopt_basepairing_y\"].append(bp_y)\n",
        "\n",
        "        self[\"subopt_NumStructs\"] = counter\n",
        "\n",
        "        self._cleanup()\n",
        "        self[\"program\"] = \"subopt\"\n",
        "\n",
        "        #print \"Minimum free energy and suboptimal secondary structures have been calculated.\"\n",
        "\n",
        "    def energy(self, strands, base_pairing_x, base_pairing_y, Temp = 37.0, dangles = \"all\"):\n",
        "\n",
        "        self[\"energy_composition\"] = strands\n",
        "\n",
        "        if Temp <= 0: raise ValueError(\"The specified temperature must be greater than zero.\")\n",
        "\n",
        "        seq_string = \"&\".join(self[\"sequences\"])\n",
        "        strands = [len(seq) for seq in self[\"sequences\"]]\n",
        "        bracket_string = self.convert_numbered_pairs_to_bracket(strands,base_pairing_x,base_pairing_y)\n",
        "        input_string = seq_string + \"\\n\" + bracket_string + \"\\n\"\n",
        "\n",
        "        handle = open(self.prefix,\"w\")\n",
        "        handle.write(input_string)\n",
        "        handle.close()\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if dangles is \"none\":\n",
        "            dangles = \" -d0 \"\n",
        "        elif dangles is \"some\":\n",
        "            dangles = \" -d1 \"\n",
        "        elif dangles is \"all\":\n",
        "            dangles = \" -d2 \"\n",
        "\n",
        "        #Call ViennaRNA C programs\n",
        "        cmd = \"RNAeval\"\n",
        "        args = dangles + \" < \" + self.prefix\n",
        "\n",
        "        output = popen2.Popen3(cmd + args)\n",
        "\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        #if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        self[\"energy_energy\"] = []\n",
        "\n",
        "        #Skip the unnecessary output lines\n",
        "        line = output.fromchild.readline()\n",
        "\n",
        "        line = output.fromchild.readline()\n",
        "        words = line.split(\" \")\n",
        "        energy = float(words[len(words)-1].replace(\"(\",\"\").replace(\")\",\"\").replace(\"\\n\",\"\"))\n",
        "\n",
        "        self[\"program\"] = \"energy\"\n",
        "        self[\"energy_energy\"].append(energy)\n",
        "        self[\"energy_basepairing_x\"] = [base_pairing_x]\n",
        "        self[\"energy_basepairing_y\"] = [base_pairing_y]\n",
        "        self._cleanup()\n",
        "\n",
        "        return energy\n",
        "\n",
        "    def convert_bracket_to_numbered_pairs(self,bracket_string):\n",
        "\n",
        "        all_seq_len = len(bracket_string)\n",
        "        bp_x = []\n",
        "        bp_y = []\n",
        "        strands = []\n",
        "\n",
        "        for y in range(bracket_string.count(\")\")):\n",
        "            bp_y.append([])\n",
        "\n",
        "        last_nt_x_list = []\n",
        "        counter=0\n",
        "        num_strands=0\n",
        "        for (pos,letter) in enumerate(bracket_string[:]):\n",
        "            if letter is \".\":\n",
        "                counter += 1\n",
        "\n",
        "            elif letter is \"(\":\n",
        "                bp_x.append(pos-num_strands)\n",
        "                last_nt_x_list.append(pos-num_strands)\n",
        "                counter += 1\n",
        "\n",
        "            elif letter is \")\":\n",
        "                nt_x = last_nt_x_list.pop()\n",
        "                nt_x_pos = bp_x.index(nt_x)\n",
        "                bp_y[nt_x_pos] = pos-num_strands\n",
        "                counter += 1\n",
        "\n",
        "            elif letter is \"&\":\n",
        "                strands.append(counter)\n",
        "                counter=0\n",
        "                num_strands+=1\n",
        "\n",
        "            else:\n",
        "                print \"Error! Invalid character in bracket notation.\"\n",
        "\n",
        "        if len(last_nt_x_list) > 0:\n",
        "            print \"Error! Leftover unpaired nucleotides when converting from bracket notation to numbered base pairs.\"\n",
        "\n",
        "        strands.append(counter)\n",
        "        bp_x = [pos+1 for pos in bp_x[:]] #Shift so that 1st position is 1\n",
        "        bp_y = [pos+1 for pos in bp_y[:]] #Shift so that 1st position is 1\n",
        "\n",
        "        return (strands,bp_x, bp_y)\n",
        "\n",
        "    def convert_numbered_pairs_to_bracket(self,strands,bp_x,bp_y):\n",
        "\n",
        "        bp_x = [pos-1 for pos in bp_x[:]] #Shift so that 1st position is 0\n",
        "        bp_y = [pos-1 for pos in bp_y[:]] #Shift so that 1st position is 0\n",
        "\n",
        "        bracket_notation = []\n",
        "        counter=0\n",
        "        for (strand_number,seq_len) in enumerate(strands):\n",
        "            if strand_number > 0: bracket_notation.append(\"&\")\n",
        "            for pos in range(counter,seq_len+counter):\n",
        "                if pos in bp_x:\n",
        "                    bracket_notation.append(\"(\")\n",
        "                elif pos in bp_y:\n",
        "                    bracket_notation.append(\")\")\n",
        "                else:\n",
        "                    bracket_notation.append(\".\")\n",
        "            counter+=seq_len\n",
        "\n",
        "        return \"\".join(bracket_notation)\n",
        "\n",
        "    def _cleanup(self):\n",
        "\n",
        "        if os.path.exists(self.prefix): os.remove(self.prefix)\n",
        "        return\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "\n",
        "    sequences = [\"AGGGGGGATCTCCCCCCAAAAAATAAGAGGTACACATGACTAAAACTTTCAAAGGCTCAGTATTCCCACT\"] #,\"acctcctta\"]\n",
        "    test = ViennaRNA(sequences,material = \"rna37\")\n",
        "\n",
        "    test.mfe([1],Temp = 37.0, dangles = \"all\")\n",
        "\n",
        "    bp_x = test[\"mfe_basepairing_x\"][0]\n",
        "    bp_y = test[\"mfe_basepairing_y\"][0]\n",
        "    strands = test[\"totalnt\"]\n",
        "    bracket_string = test.convert_numbered_pairs_to_bracket(strands,bp_x,bp_y)\n",
        "    print bracket_string\n",
        "\n",
        "    (strands,bp_x, bp_y) = test.convert_bracket_to_numbered_pairs(bracket_string)\n",
        "\n",
        "    print \"Strands = \", strands\n",
        "    print \"bp_x = \", bp_x\n",
        "    print \"bp_y = \", bp_y\n",
        "\n",
        "    print test.energy(strands, bp_x, bp_y, dangles = \"all\")\n",
        "    test.subopt(strands,3.5,dangles = \"all\")\n",
        "    print test\n",
        "\n",
        "#    print bracket_string\n",
        "#    print test.convert_numbered_pairs_to_bracket(strands,bp_x,bp_y)\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "HJme5GBhydLv",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 131
        },
        "outputId": "3956dd29-796d-41fc-d33f-202de180268b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-1-b28961f3a4be>\"\u001b[0;36m, line \u001b[0;32m82\u001b[0m\n\u001b[0;31m    if debug == 1: print output.fromchild.read()\u001b[0m\n\u001b[0m                         ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title code of NUPACK\n",
        "#Python wrapper for NUPACK 2.0 by Dirks, Bois, Schaeffer, Winfree, and Pierce (SIAM Review)\n",
        "\n",
        "#This file is part of the Ribosome Binding Site Calculator.\n",
        "\n",
        "import os.path\n",
        "import os, popen2, time, random, string\n",
        "\n",
        "tempdir = \"/tmp\" + \"\".join([random.choice(string.digits) for x in range(6)])\n",
        "\n",
        "current_dir = os.path.dirname(os.path.abspath(__file__)) + tempdir\n",
        "if not os.path.exists(current_dir): os.mkdir(current_dir)\n",
        "\n",
        "debug=0\n",
        "\n",
        "#Class that encapsulates all of the functions from NuPACK 2.0\n",
        "class NuPACK(dict):\n",
        "\n",
        "    debug_mode = 0\n",
        "    RT = 0.61597 #gas constant times 310 Kelvin (in units of kcal/mol)\n",
        "\n",
        "    def __init__(self,Sequence_List,material):\n",
        "\n",
        "        self.ran = 0\n",
        "\n",
        "        import re\n",
        "        import string\n",
        "\n",
        "        exp = re.compile('[ATGCU]',re.IGNORECASE)\n",
        "\n",
        "        for seq in Sequence_List:\n",
        "            if exp.match(seq) == None:\n",
        "                error_string = \"Invalid letters found in inputted sequences. Only ATGCU allowed. \\n Sequence is \\\"\" + str(seq) + \"\\\".\"\n",
        "                raise ValueError(error_string)\n",
        "\n",
        "        if not material == 'rna' and not material == 'dna' and not material == \"rna1999\": raise ValueError(\"The energy model must be specified as either \"\"dna\"\", \"\"rna\"\", or \"\"rna1999\"\" .\")\n",
        "\n",
        "        self[\"sequences\"] = Sequence_List\n",
        "        self[\"material\"] = material\n",
        "\n",
        "        random.seed(time.time())\n",
        "        long_id = \"\".join([random.choice(string.letters + string.digits) for x in range(10)])\n",
        "        self.prefix = current_dir + \"/nu_temp_\" + long_id\n",
        "\n",
        "    def complexes(self,MaxStrands, Temp = 37.0, ordered = \"\", pairs = \"\", mfe = \"\", degenerate = \"\", dangles = \"some\", timeonly = \"\", quiet=\"\", AdditionalComplexes = []):\n",
        "        \"\"\"A wrapper for the complexes command, which calculates the equilibrium probability of the formation of a multi-strand\n",
        "        RNA or DNA complex with a user-defined maximum number of strands.  Additional complexes may also be included by the user.\"\"\"\n",
        "\n",
        "        if Temp <= 0: raise ValueError(\"The specified temperature must be greater than zero.\")\n",
        "        if int(MaxStrands) <= 0: raise ValueError(\"The maximum number of strands must be greater than zero.\")\n",
        "\n",
        "        #Write input files\n",
        "        self._write_input_complexes(MaxStrands, AdditionalComplexes)\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if ordered: ordered = \" -ordered \"\n",
        "        if pairs: pairs = \" -pairs \"\n",
        "        if mfe: mfe = \" -mfe \"\n",
        "        if degenerate: degenerate = \" -degenerate \"\n",
        "        if timeonly: timeonly = \" -timeonly \"\n",
        "        if quiet: quiet = \" -quiet \"\n",
        "        dangles = \"-dangles \" + dangles + \" \"\n",
        "\n",
        "\n",
        "        #Call NuPACK C programs\n",
        "        cmd = \"complexes\"\n",
        "        args = \" -T \" + str(Temp) + \" -material \" + material + \" \" + ordered + pairs + mfe + degenerate \\\n",
        "        + dangles + timeonly + quiet + \" \"\n",
        "\n",
        "        output = popen2.Popen3(cmd + args + self.prefix)\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        #Read output files\n",
        "        self._read_output_cx()\n",
        "        self._cleanup(\"cx\")\n",
        "\n",
        "\n",
        "        if ordered:\n",
        "            self._read_output_ocx()\n",
        "            self._read_output_ocx_mfe()\n",
        "\n",
        "            self._cleanup(\"ocx\")\n",
        "            self._cleanup(\"ocx-mfe\")\n",
        "            self._cleanup(\"ocx-key\")\n",
        "        self._cleanup(\"in\")\n",
        "\n",
        "        #print \"Complex energies and secondary structures calculated.\"\n",
        "        self.ran = 1\n",
        "        self[\"program\"] = \"complexes\"\n",
        "\n",
        "    def mfe(self, strands,Temp = 37.0, multi = \" -multi\", pseudo = \"\", degenerate = \"\", dangles = \"some\"):\n",
        "\n",
        "        self[\"mfe_composition\"] = strands\n",
        "\n",
        "        if Temp <= 0: raise ValueError(\"The specified temperature must be greater than zero.\")\n",
        "\n",
        "        if (multi == 1 and pseudo == 1): raise ValueError(\"The pseudoknot algorithm does not work with the -multi option.\")\n",
        "\n",
        "        #Write input files\n",
        "        self._write_input_mfe(strands)\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if multi == \"\": multi = \"\"\n",
        "        if pseudo: pseudo = \" -pseudo\"\n",
        "        if degenerate: degenerate = \" -degenerate \"\n",
        "        dangles = \" -dangles \" + dangles + \" \"\n",
        "\n",
        "        #Call NuPACK C programs\n",
        "        cmd = \"mfe\"\n",
        "        args = \" -T \" + str(Temp) + multi + pseudo + \" -material \" + material + degenerate + dangles + \" \"\n",
        "\n",
        "        output = popen2.Popen3(cmd + args + self.prefix)\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        self._read_output_mfe()\n",
        "        self._cleanup(\"mfe\")\n",
        "        self._cleanup(\"in\")\n",
        "        self[\"program\"] = \"mfe\"\n",
        "        #print \"Minimum free energy secondary structure has been calculated.\"\n",
        "\n",
        "\n",
        "    def subopt(self, strands,energy_gap,Temp = 37.0, multi = \" -multi\", pseudo = \"\", degenerate = \"\", dangles = \"some\"):\n",
        "\n",
        "        self[\"subopt_composition\"] = strands\n",
        "\n",
        "        if Temp <= 0: raise ValueError(\"The specified temperature must be greater than zero.\")\n",
        "\n",
        "        if (multi == 1 and pseudo == 1): raise ValueError(\"The pseudoknot algorithm does not work with the -multi option.\")\n",
        "\n",
        "        #Write input files\n",
        "        self._write_input_subopt(strands,energy_gap)\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if multi == \"\": multi = \"\"\n",
        "        if pseudo: pseudo = \" -pseudo\"\n",
        "        if degenerate: degenerate = \" -degenerate \"\n",
        "        dangles = \" -dangles \" + dangles + \" \"\n",
        "\n",
        "        #Call NuPACK C programs\n",
        "        cmd = \"subopt\"\n",
        "        args = \" -T \" + str(Temp) + multi + pseudo + \" -material \" + material + degenerate + dangles + \" \"\n",
        "\n",
        "        output = popen2.Popen3(cmd + args + self.prefix)\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        self._read_output_subopt()\n",
        "        self._cleanup(\"subopt\")\n",
        "        self._cleanup(\"in\")\n",
        "        self[\"program\"] = \"subopt\"\n",
        "\n",
        "        #print \"Minimum free energy and suboptimal secondary structures have been calculated.\"\n",
        "\n",
        "    def energy(self, strands, base_pairing_x, base_pairing_y, Temp = 37.0, multi = \" -multi\", pseudo = \"\", degenerate = \"\", dangles = \"some\"):\n",
        "\n",
        "        self[\"energy_composition\"] = strands\n",
        "\n",
        "        if Temp <= 0: raise ValueError(\"The specified temperature must be greater than zero.\")\n",
        "\n",
        "        if (multi == 1 and pseudo == 1): raise ValueError(\"The pseudoknot algorithm does not work with the -multi option.\")\n",
        "\n",
        "        #Write input files\n",
        "        self._write_input_energy(strands,base_pairing_x,base_pairing_y)\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if multi == \"\": multi = \"\"\n",
        "        if pseudo: pseudo = \" -pseudo\"\n",
        "        if degenerate: degenerate = \" -degenerate \"\n",
        "        dangles = \" -dangles \" + dangles + \" \"\n",
        "\n",
        "        #Call NuPACK C programs\n",
        "        cmd = \"energy\"\n",
        "        args = \" -T \" + str(Temp) + multi + pseudo + \" -material \" + material + degenerate + dangles + \" \"\n",
        "\n",
        "        output = popen2.Popen3(cmd + args + self.prefix)\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        #if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        self[\"energy_energy\"] = []\n",
        "\n",
        "        #Skip the comments of the text file\n",
        "        line = output.fromchild.readline()\n",
        "        while line[0]==\"%\":\n",
        "            line = output.fromchild.readline()\n",
        "\n",
        "\n",
        "        energy = float(line)\n",
        "        self[\"program\"] = \"energy\"\n",
        "        self[\"energy_energy\"].append(energy)\n",
        "        self[\"energy_basepairing_x\"] = [base_pairing_x]\n",
        "        self[\"energy_basepairing_y\"] = [base_pairing_y]\n",
        "        self._cleanup(\"in\")\n",
        "\n",
        "        return energy\n",
        "\n",
        "    def pfunc(self, strands, Temp = 37.0, multi = \" -multi\", pseudo = \"\", degenerate = \"\", dangles = \"some\"):\n",
        "\n",
        "        self[\"pfunc_composition\"] = strands\n",
        "\n",
        "        if Temp <= 0: raise ValueError(\"The specified temperature must be greater than zero.\")\n",
        "\n",
        "        if (multi == 1 and pseudo == 1): raise ValueError(\"The pseudoknot algorithm does not work with the -multi option.\")\n",
        "\n",
        "        #Write input files\n",
        "        #Input for pfunc is the same as mfe\n",
        "        self._write_input_mfe(strands)\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if multi == \"\": multi = \"\"\n",
        "        if pseudo: pseudo = \" -pseudo\"\n",
        "        if degenerate: degenerate = \" -degenerate \"\n",
        "        dangles = \" -dangles \" + dangles + \" \"\n",
        "\n",
        "        #Call NuPACK C programs\n",
        "        cmd = \"pfunc\"\n",
        "        args = \" -T \" + str(Temp) + multi + pseudo + \" -material \" + material + degenerate + dangles + \" \"\n",
        "\n",
        "        output = popen2.Popen3(cmd + args + self.prefix)\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        #if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        #Skip the comments of the text file\n",
        "        line = output.fromchild.readline()\n",
        "        words = line.split(\" \")\n",
        "        while line[0]==\"%\" or words[0] == \"Attempting\":\n",
        "            line = output.fromchild.readline()\n",
        "            words = line.split(\" \")\n",
        "\n",
        "        energy = float(line)\n",
        "\n",
        "        line = output.fromchild.readline()\n",
        "        partition_function = float(line)\n",
        "\n",
        "        self[\"program\"] = \"pfunc\"\n",
        "        self[\"pfunc_energy\"] = energy\n",
        "        self[\"pfunc_partition_function\"] = partition_function\n",
        "        self._cleanup(\"in\")\n",
        "\n",
        "        return partition_function\n",
        "\n",
        "    def count(self, strands, Temp = 37.0, multi = \" -multi\", pseudo = \"\", degenerate = \"\", dangles = \"some\"):\n",
        "\n",
        "        self[\"count_composition\"] = strands\n",
        "\n",
        "        if (multi == 1 and pseudo == 1): raise ValueError(\"The pseudoknot algorithm does not work with the -multi option.\")\n",
        "\n",
        "        #Write input files\n",
        "        #Input for count is the same as mfe\n",
        "        self._write_input_mfe(strands)\n",
        "\n",
        "        #Set arguments\n",
        "        material = self[\"material\"]\n",
        "        if multi == \"\": multi = \"\"\n",
        "        if pseudo: pseudo = \" -pseudo\"\n",
        "        if degenerate: degenerate = \" -degenerate \"\n",
        "        dangles = \" -dangles \" + dangles + \" \"\n",
        "\n",
        "        #Call NuPACK C programs\n",
        "        cmd = \"count\"\n",
        "        args = \" -T \" + str(Temp) + multi + pseudo + \" -material \" + material + degenerate + dangles + \" \"\n",
        "\n",
        "        output = popen2.Popen3(cmd + args + self.prefix)\n",
        "        while output.poll() < 0:\n",
        "            try:\n",
        "                output.wait()\n",
        "                time.sleep(0.001)\n",
        "            except:\n",
        "                break\n",
        "\n",
        "        #if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        #Skip the comments of the text file\n",
        "        line = output.fromchild.readline()\n",
        "        words = line.split(\" \")\n",
        "        while line[0]==\"%\" or words[0] == \"Attempting\":\n",
        "            line = output.fromchild.readline()\n",
        "            words = line.split(\" \")\n",
        "\n",
        "        number = float(line)\n",
        "\n",
        "        self[\"program\"] = \"count\"\n",
        "        self[\"count_number\"] = number\n",
        "        self._cleanup(\"in\")\n",
        "\n",
        "        return number\n",
        "\n",
        "    def _write_input_energy(self,strands,base_pairing_x,base_pairing_y):\n",
        "        #Creates the input file for energy NUPACK functions\n",
        "        #strands is a list containing the number of each strand in the complex (assumes -multi flag is used)\n",
        "        #base_pairing_x and base_pairing_y is a list of base pairings of the strands s.t. #x < #y are base paired\n",
        "\n",
        "        NumStrands = len(self[\"sequences\"])\n",
        "        input_str = str(NumStrands) + \"\\n\"\n",
        "        for seq in self[\"sequences\"]:\n",
        "            input_str = input_str + seq + \"\\n\"\n",
        "\n",
        "        NumEachStrands = \"\"\n",
        "        for num in strands:\n",
        "            NumEachStrands = NumEachStrands + str(num) + \" \"\n",
        "\n",
        "        input_str = input_str + NumEachStrands + \"\\n\"\n",
        "        for pos in range(len(base_pairing_x)):\n",
        "            input_str = input_str + str(base_pairing_x[pos]) + \"\\t\" + str(base_pairing_y[pos]) + \"\\n\"\n",
        "\n",
        "        handle = open(self.prefix + \".in\", \"w\")\n",
        "        handle.writelines(input_str)\n",
        "        handle.close()\n",
        "\n",
        "    def _write_input_subopt(self,strands,energy_gap):\n",
        "        #Creates the input file for mfe and subopt NUPACK functions\n",
        "        #strands is a list containing the number of each strand in the complex (assumes -multi flag is used)\n",
        "\n",
        "        NumStrands = len(self[\"sequences\"])\n",
        "        input_str = str(NumStrands) + \"\\n\"\n",
        "        for seq in self[\"sequences\"]:\n",
        "            input_str = input_str + seq + \"\\n\"\n",
        "\n",
        "        NumEachStrands = \"\"\n",
        "        for num in strands:\n",
        "            NumEachStrands = NumEachStrands + str(num) + \" \"\n",
        "\n",
        "        input_str = input_str + NumEachStrands + \"\\n\"\n",
        "        input_str = input_str + str(energy_gap) + \"\\n\"\n",
        "\n",
        "        handle = open(self.prefix + \".in\", \"w\")\n",
        "        handle.writelines(input_str)\n",
        "        handle.close()\n",
        "\n",
        "    def _write_input_mfe(self,strands):\n",
        "        #Creates the input file for mfe and subopt NUPACK functions\n",
        "        #strands is a list containing the number of each strand in the complex (assumes -multi flag is used)\n",
        "\n",
        "        NumStrands = len(self[\"sequences\"])\n",
        "        input_str = str(NumStrands) + \"\\n\"\n",
        "        for seq in self[\"sequences\"]:\n",
        "            input_str = input_str + seq + \"\\n\"\n",
        "\n",
        "        NumEachStrands = \"\"\n",
        "        for num in strands:\n",
        "            NumEachStrands = NumEachStrands + str(num) + \" \"\n",
        "\n",
        "        input_str = input_str + NumEachStrands + \"\\n\"\n",
        "\n",
        "        handle = open(self.prefix + \".in\", \"w\")\n",
        "        handle.writelines(input_str)\n",
        "        handle.close()\n",
        "\n",
        "    def _write_input_complexes(self, MaxStrands, AdditionalComplexes = [] ):\n",
        "\n",
        "        #First, create the input string for file.in to send into NUPACK\n",
        "        NumStrands = len(self[\"sequences\"])\n",
        "        input_str = str(NumStrands) + \"\\n\"\n",
        "        for seq in self[\"sequences\"]:\n",
        "            input_str = input_str + seq + \"\\n\"\n",
        "        input_str = input_str + str(MaxStrands) + \"\\n\"\n",
        "\n",
        "        handle = open(self.prefix + \".in\", \"w\")\n",
        "        handle.writelines(input_str)\n",
        "        handle.close()\n",
        "\n",
        "        if len(AdditionalComplexes) > 0:\n",
        "            #The user may also specify additional complexes composed of more than MaxStrands strands. Create the input string detailing this.\n",
        "            counter=0\n",
        "            counts = [[]]\n",
        "            added=[]\n",
        "            for (complexes,i) in zip(AdditionalComplexes,range(len(AdditionalComplexes))):\n",
        "\n",
        "                if len(complexes) <= MaxStrands: #Remove complexes if they have less than MaxStrands strands.\n",
        "                    AdditionalComplexes.pop(i)\n",
        "                else:\n",
        "                    counts.append([])\n",
        "                    added.append(0)\n",
        "                    for j in range(NumStrands): #Count the number of each unique strand in each complex and save it to counts\n",
        "                        counts[counter].append(complexes.count(j+1))\n",
        "                    counter += 1\n",
        "\n",
        "            list_str = \"\"\n",
        "            for i in range(len(counts)-1):\n",
        "                if added[i] == 0:\n",
        "                    list_str = list_str + \"C \" + \" \".join([str(count) for count in counts[i]]) + \"\\n\"\n",
        "                    list_str = list_str + \" \".join([str(strand) for strand in AdditionalComplexes[i]]) + \"\\n\"\n",
        "                    added[i] = 1\n",
        "                    for j in range(i+1,len(counts)-1):\n",
        "                        if counts[i] == counts[j] and added[j] == 0:\n",
        "                            list_str = list_str + \" \".join([str(strand) for strand in AdditionalComplexes[j]]) + \"\\n\"\n",
        "                            added[j]=1\n",
        "\n",
        "            handle = open(self.prefix + \".list\", \"w\")\n",
        "            handle.writelines(list_str)\n",
        "            handle.close()\n",
        "\n",
        "    def _read_output_cx(self):\n",
        "\n",
        "        #Read the prefix.cx output text file generated by NuPACK and write its data to instanced attributes\n",
        "    #Output: energies of unordered complexes in key \"unordered_energies\"\n",
        "    #Output: strand composition of unordered complexes in key \"unordered_complexes\"\n",
        "\n",
        "        handle = open(self.prefix+\".cx\", \"rU\")\n",
        "\n",
        "        line = handle.readline()\n",
        "\n",
        "        #Read some useful data from the comments of the text file\n",
        "        while line[0]==\"%\":\n",
        "\n",
        "            words=line.split()\n",
        "\n",
        "            if len(words) > 7 and words[1] == \"Number\" and words[2] == \"of\" and words[3] == \"complexes\" and words[4] == \"from\" and words[5] == \"enumeration:\":\n",
        "                self[\"numcomplexes\"] =int(words[6])\n",
        "\n",
        "            elif len(words) > 8 and words[1] == \"Total\" and words[2] == \"number\" and words[3] == \"of\" and words[4] ==\"permutations\" and words[5] == \"to\" and words[6] == \"calculate:\":\n",
        "                self[\"num_permutations\"] = int(words[7])\n",
        "\n",
        "            line = handle.readline()\n",
        "\n",
        "        self[\"unordered_energies\"] = []\n",
        "        self[\"unordered_complexes\"] = []\n",
        "        self[\"unordered_composition\"] = []\n",
        "\n",
        "        while line:\n",
        "            words=line.split()\n",
        "\n",
        "            if not words[0] == \"%\":\n",
        "\n",
        "                complex = words[0]\n",
        "                strand_compos = [int(f) for f in words[1:len(words)-1]]\n",
        "                energy = float(words[len(words)-1])\n",
        "\n",
        "                self[\"unordered_complexes\"].append(complex)\n",
        "                self[\"unordered_energies\"].append(energy)\n",
        "                self[\"unordered_composition\"].append(strand_compos)\n",
        "\n",
        "            line = handle.readline()\n",
        "        handle.close()\n",
        "\n",
        "    def _read_output_ocx(self):\n",
        "\n",
        "    #Read the prefix.ocx output text file generated by NuPACK and write its data to instanced attributes\n",
        "    #Output: energies of ordered complexes in key \"ordered_energies\"\n",
        "    #Output: number of permutations and strand composition of ordered complexes in key \"ordered_complexes\"\n",
        "\n",
        "        handle = open(self.prefix+\".ocx\", \"rU\")\n",
        "\n",
        "        line = handle.readline()\n",
        "\n",
        "        #Read some useful data from the comments of the text file\n",
        "        while line[0]==\"%\":\n",
        "\n",
        "            words=line.split()\n",
        "\n",
        "            if len(words) > 7 and words[1] == \"Number\" and words[2] == \"of\" and words[3] == \"complexes\" and words[4] == \"from\" and words[5] == \"enumeration:\":\n",
        "                self[\"numcomplexes\"] =int(words[6])\n",
        "\n",
        "            elif len(words) > 8 and words[1] == \"Total\" and words[2] == \"number\" and words[3] == \"of\" and words[4] ==\"permutations\" and words[5] == \"to\" and words[6] == \"calculate:\":\n",
        "                self[\"num_permutations\"] = int(words[7])\n",
        "\n",
        "            line = handle.readline()\n",
        "\n",
        "        self[\"ordered_complexes\"] = []\n",
        "        self[\"ordered_energies\"] = []\n",
        "        self[\"ordered_permutations\"] = []\n",
        "        self[\"ordered_composition\"] = []\n",
        "\n",
        "\n",
        "        while line:\n",
        "            words=line.split()\n",
        "\n",
        "            if not words[0] == \"%\":\n",
        "\n",
        "                complex = words[0]\n",
        "                permutations = words[1]\n",
        "                strand_compos = [int(f) for f in words[2:len(words)-1]]\n",
        "                energy = float(words[len(words)-1])\n",
        "\n",
        "                self[\"ordered_complexes\"].append(complex)\n",
        "                self[\"ordered_permutations\"].append(permutations)\n",
        "                self[\"ordered_energies\"].append(energy)\n",
        "                self[\"ordered_composition\"].append(strand_compos)\n",
        "\n",
        "            line = handle.readline()\n",
        "        handle.close()\n",
        "\n",
        "    def _read_output_ocx_mfe(self):\n",
        "    #Read the prefix.ocx output text file generated by NuPACK and write its data to instanced attributes\n",
        "    #Output: energy of mfe of each complex in key \"ordered_energy\"\n",
        "\n",
        "\n",
        "        #Make sure that the ocx file has already been read.\n",
        "        if not (self.has_key(\"ordered_complexes\") and self.has_key(\"ordered_permutations\") and self.has_key(\"ordered_energies\") and self.has_key(\"ordered_composition\")):\n",
        "            self._read_output_ocx(self,prefix)\n",
        "\n",
        "        handle = open(self.prefix+\".ocx-mfe\", \"rU\")\n",
        "\n",
        "        #Skip the comments of the text file\n",
        "        line = handle.readline()\n",
        "        while line[0]==\"%\":\n",
        "            line = handle.readline()\n",
        "\n",
        "        self[\"ordered_basepairing_x\"] = []\n",
        "        self[\"ordered_basepairing_y\"] = []\n",
        "        self[\"ordered_energy\"] = []\n",
        "        self[\"ordered_totalnt\"]=[]\n",
        "\n",
        "        while line:\n",
        "            words=line.split()\n",
        "\n",
        "            if not line == \"\\n\" and not words[0] == \"%\" and not words[0] == \"\":\n",
        "\n",
        "                #Read the line containing the number of total nucleotides in the complex\n",
        "                totalnt = words[0]\n",
        "\n",
        "                self[\"ordered_totalnt\"].append(totalnt)\n",
        "\n",
        "                #Read the line containing the mfe\n",
        "                words = handle.readline().split()\n",
        "                mfe = float(words[0])\n",
        "\n",
        "                self[\"ordered_energy\"].append(mfe)\n",
        "\n",
        "                #Skip the line containing the dot/parens description of the secondary structure\n",
        "                line = handle.readline()\n",
        "\n",
        "                #Read in the lines containing the base pairing description of the secondary structure\n",
        "                #Continue reading until a % comment\n",
        "                bp_x = []\n",
        "                bp_y = []\n",
        "\n",
        "                line = handle.readline()\n",
        "                words = line.split()\n",
        "                while not line == \"\\n\" and not words[0] == \"%\":\n",
        "                    bp_x.append(int(words[0]))\n",
        "                    bp_y.append(int(words[1]))\n",
        "                    words = handle.readline().split()\n",
        "\n",
        "                self[\"ordered_basepairing_x\"].append(bp_x)\n",
        "                self[\"ordered_basepairing_y\"].append(bp_y)\n",
        "\n",
        "            line = handle.readline()\n",
        "        handle.close()\n",
        "\n",
        "    def _read_output_mfe(self):\n",
        "    #Read the prefix.mfe output text file generated by NuPACK and write its data to instanced attributes\n",
        "    #Output: total sequence length and minimum free energy\n",
        "    #Output: list of base pairings describing the secondary structure\n",
        "\n",
        "        handle = open(self.prefix+\".mfe\", \"rU\")\n",
        "\n",
        "        #Skip the comments of the text file\n",
        "        line = handle.readline()\n",
        "        while line[0]==\"%\":\n",
        "            line = handle.readline()\n",
        "\n",
        "        self[\"mfe_basepairing_x\"] = []\n",
        "        self[\"mfe_basepairing_y\"] = []\n",
        "        self[\"mfe_energy\"] = []\n",
        "        self[\"totalnt\"]=[]\n",
        "\n",
        "        counter = 0\n",
        "        while line:\n",
        "            words=line.split()\n",
        "\n",
        "            if not line == \"\\n\" and not words[0] == \"%\" and not words[0] == \"\":\n",
        "\n",
        "                #Read the line containing the number of total nucleotides in the complex\n",
        "                totalnt = words[0]\n",
        "\n",
        "                self[\"totalnt\"].append(totalnt)\n",
        "                counter += 1\n",
        "\n",
        "                #Read the line containing the mfe\n",
        "                words = handle.readline().split()\n",
        "                mfe = float(words[0])\n",
        "\n",
        "                self[\"mfe_energy\"].append(mfe)\n",
        "\n",
        "                #Skip the line containing the dot/parens description of the secondary structure\n",
        "                line = handle.readline()\n",
        "\n",
        "                #Read in the lines containing the base pairing description of the secondary structure\n",
        "                #Continue reading until a % comment\n",
        "                bp_x = []\n",
        "                bp_y = []\n",
        "\n",
        "                line = handle.readline()\n",
        "                words = line.split()\n",
        "                while not line == \"\\n\" and not words[0] == \"%\":\n",
        "                    bp_x.append(int(words[0]))\n",
        "                    bp_y.append(int(words[1]))\n",
        "                    words = handle.readline().split()\n",
        "\n",
        "                self[\"mfe_basepairing_x\"].append(bp_x)\n",
        "                self[\"mfe_basepairing_y\"].append(bp_y)\n",
        "\n",
        "            line = handle.readline()\n",
        "        handle.close()\n",
        "\n",
        "        self[\"mfe_NumStructs\"] = counter\n",
        "\n",
        "    def _read_output_subopt(self):\n",
        "    #Read the prefix.subopt output text file generated by NuPACK and write its data to instanced attributes\n",
        "    #Output: total sequence length and minimum free energy\n",
        "    #Output: list of base pairings describing the secondary structure\n",
        "\n",
        "        handle = open(self.prefix+\".subopt\", \"rU\")\n",
        "\n",
        "        #Skip the comments of the text file\n",
        "        line = handle.readline()\n",
        "        while line[0]==\"%\":\n",
        "            line = handle.readline()\n",
        "\n",
        "        self[\"subopt_basepairing_x\"] = []\n",
        "        self[\"subopt_basepairing_y\"] = []\n",
        "        self[\"subopt_energy\"] = []\n",
        "        self[\"totalnt\"]=[]\n",
        "\n",
        "        counter=0\n",
        "\n",
        "        while line:\n",
        "            words=line.split()\n",
        "\n",
        "            if not line == \"\\n\" and not words[0] == \"%\" and not words[0] == \"\":\n",
        "\n",
        "                #Read the line containing the number of total nucleotides in the complex\n",
        "                totalnt = words[0]\n",
        "\n",
        "                self[\"totalnt\"].append(totalnt)\n",
        "                counter += 1\n",
        "\n",
        "                #Read the line containing the mfe\n",
        "                words = handle.readline().split()\n",
        "                mfe = float(words[0])\n",
        "\n",
        "                self[\"subopt_energy\"].append(mfe)\n",
        "\n",
        "                #Skip the line containing the dot/parens description of the secondary structure\n",
        "                line = handle.readline()\n",
        "\n",
        "                #Read in the lines containing the base pairing description of the secondary structure\n",
        "                #Continue reading until a % comment\n",
        "                bp_x = []\n",
        "                bp_y = []\n",
        "\n",
        "                line = handle.readline()\n",
        "                words = line.split()\n",
        "                while not line == \"\\n\" and not words[0] == \"%\":\n",
        "                    bp_x.append(int(words[0]))\n",
        "                    bp_y.append(int(words[1]))\n",
        "                    words = handle.readline().split()\n",
        "\n",
        "                self[\"subopt_basepairing_x\"].append(bp_x)\n",
        "                self[\"subopt_basepairing_y\"].append(bp_y)\n",
        "\n",
        "            line = handle.readline()\n",
        "        handle.close()\n",
        "\n",
        "        self[\"subopt_NumStructs\"] = counter\n",
        "\n",
        "    def _cleanup(self,suffix):\n",
        "\n",
        "        if os.path.exists(self.prefix+\".\"+suffix): os.remove(self.prefix+\".\"+suffix)\n",
        "\n",
        "        return\n",
        "\n",
        "    def export_PDF(self, complex_ID, name = \"\", filename = \"temp.pdf\", program = None):\n",
        "        \"\"\"Uses Zuker's sir_graph_ng and ps2pdf.exe to convert a secondary structure described in .ct format\n",
        "        to a PDF of the RNA\"\"\"\n",
        "\n",
        "        if program is None:\n",
        "            program = self[\"program\"]\n",
        "\n",
        "        inputfile = \"temp.ct\"\n",
        "        self.Convert_to_ct(complex_ID,name,inputfile,program)\n",
        "\n",
        "\n",
        "        cmd = \"sir_graph_ng\" #Assumes it's on the path\n",
        "        args = \"-p\" #to PostScript file\n",
        "        output = popen2.Popen3(cmd + \" \" + args + \" \" + inputfile,\"r\")\n",
        "        output.wait()\n",
        "        if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        inputfile = inputfile[0:len(inputfile)-2] + \"ps\"\n",
        "\n",
        "        cmd = \"ps2pdf\" #Assumes it's on the path\n",
        "        output = popen2.Popen3(cmd + \" \" + inputfile,\"r\")\n",
        "        output.wait()\n",
        "        if debug == 1: print output.fromchild.read()\n",
        "\n",
        "        outputfile = inputfile[0:len(inputfile)-2] + \"pdf\"\n",
        "\n",
        "        #Remove the temporary file \"temp.ct\" if it exists\n",
        "        if os.path.exists(\"temp.ct\"): os.remove(\"temp.ct\")\n",
        "\n",
        "        #Remove the temporary Postscript file if it exists\n",
        "        if os.path.exists(inputfile): os.remove(inputfile)\n",
        "\n",
        "        #Rename the output file to the desired filename.\n",
        "        if os.path.exists(outputfile): os.rename(outputfile,filename)\n",
        "        #Done!\n",
        "\n",
        "    def Convert_to_ct(self,complex_ID,name,filename = \"temp.ct\",program = \"ordered\"):\n",
        "        \"\"\"Converts the secondary structure of a single complex into the .ct file format, which is used\n",
        "        with sir_graph_ng (or other programs) to create an image of the secondary structure.\"\"\"\n",
        "\n",
        "        #hacksy way of reading from data produced by 'complex', by 'mfe', or by 'subopt'\n",
        "        data_x = program + \"_basepairing_x\"\n",
        "        data_y = program + \"_basepairing_y\"\n",
        "        mfe_name = program + \"_energy\"\n",
        "        composition_name = program + \"_composition\"\n",
        "\n",
        "        #Format of .ct file\n",
        "\n",
        "        #Header: <Total # nt> \\t dG = <# mfe> kcal/mol \\t <name of sequence>\n",
        "        #The Rest:\n",
        "        #<nt num> \\t <bp letter> \\t <3' neighbor> \\t <5' neighbor> \\t <# of bp'ing, 0 if none> \\t ...\n",
        "        #<strand-specific nt num> \\t <3' neighbor if connected by helix> \\t <5' neighbor if connected by helix>\n",
        "\n",
        "        #Extract the data for the desired complex using complex_ID\n",
        "        bp_x = self[data_x][complex_ID]\n",
        "        bp_y = self[data_y][complex_ID]\n",
        "        mfe = self[mfe_name][complex_ID]\n",
        "\n",
        "        if program == \"mfe\" or program == \"subopt\" or program == \"energy\":\n",
        "            composition = self[composition_name]\n",
        "        elif program == \"ordered\" or program == \"unordered\":\n",
        "            composition = self[composition_name][complex_ID]\n",
        "\n",
        "\n",
        "        #Determine concatenated sequence of all strands, their beginnings, and ends\n",
        "        allseq = \"\"\n",
        "        strand_begins = []\n",
        "        strand_ends = []\n",
        "\n",
        "        #Seemingly, the format of the composition is different for the program complex vs. mfe/subopt\n",
        "        #for mfe/subopt, the composition is the list of strand ids\n",
        "        #for complex, it is the number of each strand (in strand id order) in the complex\n",
        "        #for mfe/subopt, '1 2 2 3' refers to 1 strand of 1, 2 strands of 2, and 1 strand of 3.\n",
        "        #for complex, '1 2 2 3' refers to 1 strand of 1, 2 strands of 2, 2 strands of 3, and 3 strands of 4'.\n",
        "        #what a mess.\n",
        "\n",
        "        if program == \"mfe\" or program == \"subopt\" or program == \"energy\":\n",
        "            for strand_id in composition:\n",
        "                strand_begins.append(len(allseq) + 1)\n",
        "                allseq = allseq + self[\"sequences\"][strand_id-1]\n",
        "                strand_ends.append(len(allseq))\n",
        "\n",
        "        else:\n",
        "            for (num_strands,strand_id) in zip(composition,range(len(composition))):\n",
        "                for j in range(num_strands):\n",
        "                    strand_begins.append(len(allseq) + 1)\n",
        "                    allseq = allseq + self[\"sequences\"][strand_id]\n",
        "                    strand_ends.append(len(allseq))\n",
        "\n",
        "        seq_len = len(allseq)\n",
        "\n",
        "        #print \"Seq Len = \", seq_len, \"  Composition = \", composition\n",
        "        #print \"Sequence = \", allseq\n",
        "        #print \"Base pairing (x) = \", bp_x\n",
        "        #print \"Base pairing (y) = \", bp_y\n",
        "\n",
        "\n",
        "        #Create the header\n",
        "        header = str(seq_len) + \"\\t\" + \"dG = \" + str(mfe) + \" kcal/mol\" + \"\\t\" + name + \"\\n\"\n",
        "\n",
        "        #Open the file\n",
        "        handle = open(filename,\"w\")\n",
        "\n",
        "        #Write the header\n",
        "        handle.write(header)\n",
        "\n",
        "        #Write a line for each nt in the secondary structure\n",
        "        for i in range(1,seq_len+1):\n",
        "\n",
        "\n",
        "            for (nt,pos) in zip(strand_begins,range(len(strand_begins))):\n",
        "                if i >= nt:\n",
        "                    strand_id = pos\n",
        "\n",
        "\n",
        "            #Determine 3' and 5' neighbor\n",
        "            #If this is the beginning of a strand, then the 3' neighbor is 0\n",
        "            #If this is the end of a strand, then the 5' neighbor is 0\n",
        "\n",
        "            if i in strand_begins:\n",
        "                nb_5p = 0\n",
        "            else:\n",
        "                nb_5p = i - 1\n",
        "\n",
        "            if i in strand_ends:\n",
        "                nb_3p = 0\n",
        "            else:\n",
        "                nb_3p = i + 1\n",
        "\n",
        "\n",
        "            if i in bp_x or i in bp_y:\n",
        "                if i in bp_x: nt_bp = bp_y[bp_x.index(i)]\n",
        "                if i in bp_y: nt_bp = bp_x[bp_y.index(i)]\n",
        "            else:\n",
        "                nt_bp = 0\n",
        "\n",
        "            #Determine strand-specific counter\n",
        "            strand_counter = i - strand_begins[strand_id] + 1\n",
        "\n",
        "            #Determine the 3' and 5' neighbor helical connectivity\n",
        "            #If the ith nt is connected to its 3', 5' neighbor by a helix, then include it\n",
        "            #Otherwise, 0\n",
        "            #Helix connectivity conditions:\n",
        "            #The 5' or 3' neighbor is connected via a helix iff:\n",
        "            #a) helix start: i not bp'd, i+1 bp'd, bp_id(i+1) - 1 is bp'd, bp_id(i+1) + 1 is not bp'd\n",
        "            #b) helix end: i not bp'd, i-1 bp'd, bp_id(i-1) - 1 is not bp'd, bp_id(i-1) + 1 is bp'd\n",
        "            #c) helix continued: i and bp_id(i)+1 is bp'd, 5' helix connection is bp_id(bp_id(i)+1)\n",
        "            #d) helix continued: i and bp_id(i)-1 is bp'd, 3' helix connection is bp_id(bp_id(i)-1)\n",
        "            #Otherwise, zero.\n",
        "\n",
        "            #Init\n",
        "            hc_5p = 0\n",
        "            hc_3p = 0\n",
        "\n",
        "            if i in bp_x or i in bp_y:  #helix continued condition (c,d)\n",
        "                if i in bp_x: bp_i = bp_y[bp_x.index(i)]\n",
        "                if i in bp_y: bp_i = bp_x[bp_y.index(i)]\n",
        "\n",
        "                if bp_i+1 in bp_x or bp_i+1 in bp_y: #helix condition c\n",
        "                    if bp_i+1 in bp_x: hc_3p = bp_y[bp_x.index(bp_i+1)]\n",
        "                    if bp_i+1 in bp_y: hc_3p = bp_x[bp_y.index(bp_i+1)]\n",
        "\n",
        "                if bp_i-1 in bp_x or bp_i-1 in bp_y: #helix condition d\n",
        "                    if bp_i-1 in bp_x: hc_5p = bp_y[bp_x.index(bp_i-1)]\n",
        "                    if bp_i-1 in bp_y: hc_5p = bp_x[bp_y.index(bp_i-1)]\n",
        "\n",
        "            else: #helix start or end (a,b)\n",
        "\n",
        "                if i+1 in bp_x or i+1 in bp_y: #Start, condition a\n",
        "                    if i+1 in bp_x: bp_3p = bp_y[bp_x.index(i+1)]\n",
        "                    if i+1 in bp_y: bp_3p = bp_x[bp_y.index(i+1)]\n",
        "\n",
        "                    if bp_3p + 1 not in bp_x and bp_3p + 1 not in bp_y:\n",
        "                        hc_3p = i + 1\n",
        "\n",
        "                if i-1 in bp_x or i-1 in bp_y: #End, condition b\n",
        "                    if i-1 in bp_x: bp_5p = bp_y[bp_x.index(i-1)]\n",
        "\n",
        "                    if i-1 in bp_y: bp_5p = bp_x[bp_y.index(i-1)]\n",
        "\n",
        "                    if bp_5p - 1 not in bp_x and bp_5p - 1 not in bp_y:\n",
        "                        hc_5p = i - 1\n",
        "\n",
        "\n",
        "            line = str(i) + \"\\t\" + allseq[i-1] + \"\\t\" + str(nb_5p) + \"\\t\" + str(nb_3p) + \"\\t\" + str(nt_bp) + \"\\t\" + str(strand_counter) + \"\\t\" + str(hc_5p) + \"\\t\" + str(hc_3p) + \"\\n\"\n",
        "\n",
        "            handle.write(line)\n",
        "\n",
        "        #Close the file. Done.\n",
        "        handle.close()\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "\n",
        "    import re\n",
        "\n",
        "    #sequences = [\"AAGATTAACTTAAAAGGAAGGCCCCCCATGCGATCAGCATCAGCACTACGACTACGCGA\",\"acctcctta\",\"ACGTTGGCCTTCC\"]\n",
        "    sequences = [\"AAGATTAACTTAAAAGGAAGGCCCCCCATGCGATCAGCATCAGCACTACGACTACGCGA\"]\n",
        "\n",
        "    #Complexes\n",
        "    #Input: Max number of strands in a complex. Considers all possible combinations of strands, up to max #.\n",
        "    #'mfe': calculate mfe? 'ordered': consider ordered or unordered complexes?\n",
        "    #Other options available (see function)\n",
        "\n",
        "    AddComplexes = []\n",
        "    test = NuPACK(sequences,\"rna1999\")\n",
        "    test.complexes(3,mfe = 1, ordered=1)\n",
        "\n",
        "    print test\n",
        "\n",
        "    strand_compositions = test[\"ordered_composition\"]\n",
        "    num_complexes = len(strand_compositions)\n",
        "    num_strands = len(sequences)\n",
        "\n",
        "    for counter in range(num_complexes):\n",
        "        output = \"Complex #\" + str(counter+1) + \" composition: (\"\n",
        "        for strand_id in strand_compositions[counter][0:num_strands-1]:\n",
        "            output = output + str(strand_id) + \", \"\n",
        "        output = output + str(strand_compositions[counter][num_strands-1]) + \")\"\n",
        "\n",
        "        output = output + \"  dG (RT ln Q): \" + str(test[\"ordered_energy\"][counter]) + \" kcal/mol\"\n",
        "        output = output + \"  # Permutations: \" + str(test[\"ordered_permutations\"][counter])\n",
        "        print output\n",
        "        test.export_PDF(counter, name = \"Complex #\" + str(counter+1), filename = \"Complex_\" + str(counter) + \".pdf\", program = \"ordered\")\n",
        "\n",
        "    #Mfe\n",
        "    #Input: Number of each strand in complex.\n",
        "    #Options include RNA/DNA model, temperature, dangles, etc. (See function).\n",
        "    #Example: If there are 3 unique strands (1, 2, 3), then [1, 2, 3] is one of each strand and [1, 1, 2, 2, 3, 3] is two of each strand.\n",
        "\n",
        "    #test.mfe([1, 2], dangles = \"all\")\n",
        "    #num_complexes = test[\"mfe_NumStructs\"]  #Number of degenerate complexes (same energy)\n",
        "    #dG_mfe = test[\"mfe_energy\"]\n",
        "    #print \"There are \", num_complexes, \" configuration(s) with a minimum free energy of \", dG_mfe, \" kcal/mol.\"\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "6Rjg1Q_XzXto",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 131
        },
        "outputId": "466209b9-206b-473f-d709-d351e3b3a108"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-2-48433da823b9>\"\u001b[0;36m, line \u001b[0;32m79\u001b[0m\n\u001b[0;31m    if debug == 1: print output.fromchild.read()\u001b[0m\n\u001b[0m                         ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title the RBS calculater\n",
        "\n",
        "#Given an mRNA sequence, this Python class predicts the dG_total and translation initiation rate.\n",
        "#This file is part of the Ribosome Binding Site Calculator.\n",
        "#along with Ribosome Binding Site Calculator.  If not, see <http://www.gnu.org/licenses/>.\n",
        "\n",
        "\n",
        "from NuPACK import NuPACK\n",
        "import re\n",
        "import math\n",
        "\n",
        "class CalcError(Exception):\n",
        "    \"\"\"Base class for exceptions in this module.\"\"\"\n",
        "\n",
        "    def __init__(self, value):\n",
        "        self.value = value\n",
        "\n",
        "    def __str__(self):\n",
        "        return repr(self.value)\n",
        "\n",
        "class RBS_Calculator(NuPACK):\n",
        "\n",
        "    #From experimental characterization of ALL Predicted RBSs as of 1/20/08\n",
        "    RT_eff = 2.222\n",
        "    logK =  7.824\n",
        "    K = 2500.0\n",
        "\n",
        "    #Global parameters -- constants\n",
        "    infinity = 1e12 #For all practical purposes, here.\n",
        "    RNA_model = \"rna1999\"\n",
        "    start_codon_energies = {\"ATG\":-1.194, \"AUG\": -1.194, \"GTG\": -0.0748, \"GUG\": -0.0748, \"TTG\":-0.0435, \"UUG\": -0.0435, \"CTG\": -0.03406, \"CUG\":-0.03406} #hybridization to CAT\n",
        "    auto_dangles = True\n",
        "    dangles_default = \"all\"\n",
        "    temp = 37.0\n",
        "    optimal_spacing = 5 #aligned spacing\n",
        "\n",
        "    dG_spacing_constant_push = [12.2, 2.5, 2.0, 3.0]\n",
        "    dG_spacing_constant_pull = [0.048, 0.24, 0.0]\n",
        "    cutoff = 35 #number of nt +- start codon considering for folding\n",
        "    standby_site_length = 4 #Number of nt before SD sequence that must be unpaired for ribosome binding\n",
        "    energy_cutoff = 3.0\n",
        "    start_codons = [\"ATG\", \"AUG\", \"GTG\", \"GUG\",\"TTG\",\"UUG\"] #substituted U for T in actual calcs. Ignores CTG/CUG\n",
        "    rRNA = \"acctcctta\" #These are the last 9 nt (3' end) of the 16S rRNA in E. coli\n",
        "\n",
        "    footprint = 1000 #Footprint of the 30S complex that prevents formation of secondary structures downstream of the start codon. Here, we assume that the entire post-start RNA sequence does not form secondary structures once the 30S complex has bound.\n",
        "\n",
        "    def __init__(self, mRNA, start_range, name = \"Unnamed mRNA\", verbose = False):\n",
        "        \"\"\"Initializes the RBS Calculator class with the mRNA sequence and the range of start codon positions considered.\"\"\"\n",
        "\n",
        "        #NuPACK.__init__(self,sequences,self.RNA_model)\n",
        "\n",
        "        exp = re.compile('[ATGCU]',re.IGNORECASE)\n",
        "        if exp.match(mRNA) == None:\n",
        "            raise ValueError(\"Invalid letters found in sequence \"\"%s\"\". Only ATGCU accepted.\" % mRNA)\n",
        "\n",
        "        if start_range[0] < 0: start_range[0] = 0\n",
        "        if start_range[1] > len(mRNA): start_range[1] = len(mRNA)\n",
        "\n",
        "        self.name = name\n",
        "        self.mRNA_input = mRNA.upper()\n",
        "        self.rRNA_len = len(self.rRNA)\n",
        "        self.mRNA_len = len(self.mRNA_input)\n",
        "        self.total_sequence_length = len(mRNA) + len(self.rRNA)\n",
        "        self.dG_rRNA = self.calc_dG_rRNA()\n",
        "        self.run = 0\n",
        "        self.start_range = start_range\n",
        "        self.verbose = verbose\n",
        "\n",
        "    def find_min(self,input_list):\n",
        "        \"\"\"Finds the minimum of a list of numbers.\"\"\"\n",
        "\n",
        "        min_item = self.infinity\n",
        "        min_index = 0\n",
        "\n",
        "        for i, item in enumerate(input_list):\n",
        "            if item < min_item:\n",
        "                min_item = item\n",
        "                min_index = i\n",
        "\n",
        "        return (min_item,min_index)\n",
        "\n",
        "\n",
        "    def find_start_codons(self,sequence):\n",
        "        \"\"\"Finds all start codons in an mRNA sequence. Creates a list.\"\"\"\n",
        "\n",
        "        self.start_position_list = []\n",
        "        self.start_codon_list = []\n",
        "\n",
        "        seq_len = len(sequence)\n",
        "        end = min(self.start_range[1],seq_len-2)\n",
        "        begin = min(self.start_range[0],end)\n",
        "\n",
        "        for i in range(begin,end+1):\n",
        "            codon = sequence[i:i+3]\n",
        "            if codon.upper() in self.start_codons:\n",
        "                self.start_position_list.append(i)\n",
        "                self.start_codon_list.append(codon)\n",
        "                yield (i,codon)\n",
        "            else:\n",
        "                pass\n",
        "\n",
        "    def calc_aligned_spacing(self,mRNA,start_pos,bp_x,bp_y):\n",
        "        \"\"\"Calculates the aligned spacing between the 16S rRNA binding site and the start codon.\"\"\"\n",
        "\n",
        "        #rRNA is the concatenated at the end of the sequence in 5' to 3' direction\n",
        "        #first: identify the farthest 3' nt in the rRNA that binds to the mRNA and return its mRNA base pairer\n",
        "\n",
        "        Ok = False\n",
        "        seq_len = len(mRNA) + self.rRNA_len\n",
        "        for (rRNA_nt) in range(seq_len,seq_len - self.rRNA_len,-1):\n",
        "\n",
        "            if rRNA_nt in bp_y:\n",
        "                rRNA_pos = bp_y.index(rRNA_nt)\n",
        "                if bp_x[rRNA_pos] < start_pos:\n",
        "                    Ok = True\n",
        "                    farthest_3_prime_rRNA = rRNA_nt - len(mRNA)\n",
        "\n",
        "                    mRNA_nt = bp_x[rRNA_pos]\n",
        "                    distance_to_start = start_pos - mRNA_nt + 1 #start_pos is counting starting from 0 (python)\n",
        "\n",
        "                    break\n",
        "                else:\n",
        "                    break\n",
        "\n",
        "        if Ok:\n",
        "            aligned_spacing = distance_to_start - farthest_3_prime_rRNA\n",
        "        else:\n",
        "            aligned_spacing = self.infinity\n",
        "\n",
        "        return aligned_spacing\n",
        "\n",
        "    def calc_dG_spacing(self, aligned_spacing):\n",
        "        \"\"\"Calculates the dG_spacing according to the value of the aligned spacing. This relationship was determined through experiments.\"\"\"\n",
        "\n",
        "        if (aligned_spacing < self.optimal_spacing):\n",
        "            ds = aligned_spacing - self.optimal_spacing\n",
        "\n",
        "            dG_spacing_penalty = self.dG_spacing_constant_push[0] / (1.0 + math.exp(self.dG_spacing_constant_push[1]*(ds + self.dG_spacing_constant_push[2] )))**self.dG_spacing_constant_push[3]\n",
        "\n",
        "        else:\n",
        "            ds = aligned_spacing - self.optimal_spacing\n",
        "            dG_spacing_penalty = self.dG_spacing_constant_pull[0] * ds * ds + self.dG_spacing_constant_pull[1] * ds + self.dG_spacing_constant_pull[2]\n",
        "\n",
        "        return dG_spacing_penalty\n",
        "\n",
        "    def calc_dG_mRNA_rRNA(self,start_pos):\n",
        "        \"\"\"Calculates the dG_mRNA_rRNA from the mRNA and rRNA sequence. Considers all feasible 16S rRNA binding sites and includes the effects of non-optimal spacing.\"\"\"\n",
        "\n",
        "        begin = max(0,start_pos-self.cutoff)\n",
        "        mRNA_len = min(len(self.mRNA_input),start_pos+self.cutoff)\n",
        "        start_pos_in_subsequence = min(start_pos, self.cutoff)\n",
        "        startpos_to_end_len = mRNA_len - start_pos_in_subsequence - begin\n",
        "\n",
        "        #1. identify a list of rRNA-binding sites. Binding sites are hybridizations between the mRNA and rRNA and can include mismatches, bulges, etc. Intra-molecular folding is also allowed within the mRNA. The subopt program is used to generate a list of optimal & suboptimal binding sites.\n",
        "        #Constraints: the entire rRNA-binding site must be upstream of the start codon\n",
        "        mRNA = self.mRNA_input[begin:start_pos]\n",
        "\n",
        "        if len(mRNA) == 0:\n",
        "            raise CalcError(\"Warning: There is a leaderless start codon, which is being ignored.\")\n",
        "\n",
        "        #print \"After exception\"\n",
        "\n",
        "        fold = NuPACK([mRNA,self.rRNA],material = self.RNA_model)\n",
        "        fold.subopt([1, 2],self.energy_cutoff,dangles = self.dangles, Temp = self.temp)\n",
        "\n",
        "        if len(fold[\"subopt_basepairing_x\"]) == 0:\n",
        "            raise CalcError(\"Warning: The 16S rRNA has no predicted binding site. Start codon is considered as leaderless and ignored.\")\n",
        "\n",
        "        #2. Calculate dG_spacing for each 16S rRNA binding site\n",
        "\n",
        "        #Calculate the aligned spacing for each binding site in the list\n",
        "        aligned_spacing = []\n",
        "        for (bp_x, bp_y) in zip(fold[\"subopt_basepairing_x\"], fold[\"subopt_basepairing_y\"]):\n",
        "            aligned_spacing.append(self.calc_aligned_spacing(mRNA, start_pos_in_subsequence, bp_x,bp_y))\n",
        "\n",
        "        dG_spacing_list = []\n",
        "        dG_mRNA_rRNA = []\n",
        "        dG_mRNA_rRNA_withspacing = []\n",
        "\n",
        "        #Calculate dG_spacing using aligned spacing value. Add it to dG_mRNA_rRNA.\n",
        "        for (counter) in range(len(fold[\"subopt_basepairing_x\"])):\n",
        "\n",
        "            dG_mRNA_rRNA.append(fold[\"subopt_energy\"][counter])\n",
        "            val = self.calc_dG_spacing(aligned_spacing[counter])\n",
        "            dG_spacing_list.append(val)\n",
        "            dG_mRNA_rRNA_withspacing.append(val + fold[\"subopt_energy\"][counter])\n",
        "\n",
        "        #3. Find 16S rRNA binding site that minimizes dG_spacing+dG_mRNA_rRNA.\n",
        "        [dG_mRNA_rRNA_folding, index] = self.find_min(dG_mRNA_rRNA_withspacing)\n",
        "        dG_spacing_final = dG_spacing_list[index]\n",
        "\n",
        "        dG_mRNA_rRNA_nospacing = dG_mRNA_rRNA[index]\n",
        "\n",
        "        #Check: Is the dG spacing large compared to the energy gap? If so, this means the list of suboptimal 16S rRNA binding sites generated by subopt is too short.\n",
        "        if dG_spacing_final > self.energy_cutoff:\n",
        "            if self.verbose: print \"Warning: The spacing penalty is greater than the energy gap. dG (spacing) = \", dG_spacing_final\n",
        "\n",
        "\n",
        "        #4. Identify the 5' and 3' ends of the identified 16S rRNA binding site. Create a base pair list.\n",
        "\n",
        "        most_5p_mRNA = self.infinity\n",
        "        most_3p_mRNA = -self.infinity\n",
        "\n",
        "        #Generate a list of rRNA-mRNA base paired nucleotides\n",
        "        bp_x_target = []\n",
        "        bp_y_target = []\n",
        "\n",
        "        bp_x = fold[\"subopt_basepairing_x\"][index]\n",
        "        bp_y = fold[\"subopt_basepairing_y\"][index]\n",
        "        for (nt_x, nt_y) in zip(bp_x, bp_y):\n",
        "            if nt_y > len(mRNA): #nt is rRNA\n",
        "                most_5p_mRNA = min(most_5p_mRNA, bp_x[bp_y.index(nt_y)])\n",
        "                most_3p_mRNA = max(most_3p_mRNA, bp_x[bp_y.index(nt_y)])\n",
        "                bp_x_target.append(nt_x)\n",
        "                bp_y_target.append(nt_y)\n",
        "\n",
        "        #The rRNA-binding site is between the nucleotides at positions most_5p_mRNA and most_3p_mRNA\n",
        "        #Now, fold the pre-sequence, rRNA-binding-sequence and post-sequence separately. Take their base pairings and combine them together. Calculate the total energy. For secondary structures, this splitting operation is allowed.\n",
        "\n",
        "        #We postulate that not all of the post-sequence can form secondary structures. Once the 30S complex binds to the mRNA, it prevents the formation of secondary structures that are mutually exclusive with ribosome binding. We define self.footprint to be the length of the 30S complex footprint. Here, we assume that the entire mRNA sequence downstream of the 16S rRNA binding site can not form secondary structures.\n",
        "\n",
        "        mRNA_pre = self.mRNA_input[begin:begin+most_5p_mRNA-1]\n",
        "        post_window_end = mRNA_len + 1\n",
        "        post_window_begin = min(start_pos + self.footprint,post_window_end) #Footprint\n",
        "        post_window_end = mRNA_len + 1\n",
        "        mRNA_post = self.mRNA_input[post_window_begin:post_window_end]\n",
        "\n",
        "        mRNA_pre_len = len(mRNA_pre)\n",
        "        mRNA_post_len = len(mRNA_post)\n",
        "        mRNA_rRNA_binding_len = most_3p_mRNA - most_5p_mRNA + 1\n",
        "        total_folded_len = mRNA_pre_len + mRNA_post_len + mRNA_rRNA_binding_len\n",
        "\n",
        "        total_bp_x = []\n",
        "        total_bp_y = []\n",
        "\n",
        "        #Calculate pre-sequence folding\n",
        "        if len(mRNA_pre) > 0:\n",
        "            fold_pre = NuPACK([mRNA_pre], material = self.RNA_model)\n",
        "            fold_pre.mfe([1], dangles = self.dangles, Temp = self.temp)\n",
        "            bp_x_pre = fold_pre[\"mfe_basepairing_x\"][0]\n",
        "            bp_y_pre = fold_pre[\"mfe_basepairing_y\"][0]\n",
        "\n",
        "        else:\n",
        "            bp_x_pre = []\n",
        "            bp_y_pre = []\n",
        "\n",
        "        #Add pre-sequence base pairings to total base pairings\n",
        "        offset = 0 #Begins at 0\n",
        "        for (nt_x, nt_y) in zip(bp_x_pre, bp_y_pre):\n",
        "            total_bp_x.append(nt_x + offset)\n",
        "            total_bp_y.append(nt_y + offset)\n",
        "\n",
        "        #Add rRNA-binding site base pairings to total base pairings\n",
        "        offset = 0 #Begins at zero\n",
        "        if startpos_to_end_len < self.cutoff:\n",
        "            rRNA_offset = startpos_to_end_len\n",
        "        else:\n",
        "            rRNA_offset = startpos_to_end_len\n",
        "\n",
        "        for (nt_x, nt_y) in zip(bp_x_target, bp_y_target):\n",
        "            total_bp_x.append(nt_x + offset)\n",
        "            total_bp_y.append(nt_y + rRNA_offset)\n",
        "\n",
        "        #Calculate post-sequence folding\n",
        "        if len(mRNA_post) > 0:\n",
        "            fold_post = NuPACK([mRNA_post], material = self.RNA_model)\n",
        "            fold_post.mfe([1], dangles = self.dangles, Temp = self.temp)\n",
        "            bp_x_post = fold_post[\"mfe_basepairing_x\"][0]\n",
        "            bp_y_post = fold_post[\"mfe_basepairing_y\"][0]\n",
        "        else:\n",
        "            bp_x_post = []\n",
        "            bp_y_post = []\n",
        "\n",
        "        offset = post_window_begin - begin\n",
        "        for (nt_x, nt_y) in zip(bp_x_post, bp_y_post):\n",
        "            total_bp_x.append(nt_x + offset)\n",
        "            total_bp_y.append(nt_y + offset)\n",
        "\n",
        "        mRNA = self.mRNA_input[begin:mRNA_len]\n",
        "        fold = NuPACK([mRNA, self.rRNA], material = self.RNA_model)\n",
        "\n",
        "        total_energy = fold.energy([1, 2], total_bp_x, total_bp_y, Temp = self.temp, dangles = self.dangles)\n",
        "\n",
        "        energy_nowindows = dG_mRNA_rRNA_nospacing\n",
        "        total_energy_withspacing = total_energy + dG_spacing_final\n",
        "\n",
        "        structure = fold\n",
        "        structure[\"program\"] = \"subopt\"\n",
        "        structure[\"mRNA\"] = mRNA\n",
        "        structure[\"MinStructureID\"] = 0\n",
        "        structure[\"dG_mRNA_rRNA\"] = total_energy\n",
        "        structure[\"dG_mRNA_rRNA_withspacing\"] = total_energy_withspacing\n",
        "        structure[\"dG_spacing\"] = dG_spacing_final\n",
        "        structure[\"subopt_energy\"] = [total_energy_withspacing]\n",
        "        structure[\"subopt_basepairing_x\"] = [total_bp_x]\n",
        "        structure[\"subopt_basepairing_y\"] = [total_bp_y]\n",
        "        structure[\"subopt_composition\"] = [1, 2]\n",
        "        structure[\"bp_x\"] = total_bp_x\n",
        "        structure[\"bp_y\"] = total_bp_y\n",
        "\n",
        "        return (total_energy_withspacing, structure)\n",
        "\n",
        "    def calc_dG_standby_site(self,structure_old, rRNA_binding = True):\n",
        "        \"\"\"Calculates the dG_standby given the structure of the mRNA:rRNA complex\"\"\"\n",
        "\n",
        "        #To calculate the mfe structure while disallowing base pairing at the standby site, we split the folded mRNA sequence into three parts: (i) a pre-sequence (before the standby site) that can fold; (ii) the standby site, which can not fold; (iii) the 16S rRNA binding site and downstream sequence, which has been previously folded.\n",
        "\n",
        "        import copy\n",
        "        structure = copy.deepcopy(structure_old)\n",
        "        mRNA = structure[\"mRNA\"]\n",
        "        bp_x = structure[\"bp_x\"]\n",
        "        bp_y = structure[\"bp_y\"]\n",
        "        energy_before = structure[\"dG_mRNA_rRNA\"] #without spacing effects\n",
        "\n",
        "        #Identify the most 5p mRNA nt that is bound to rRNA\n",
        "        for (nt_x, nt_y) in zip(bp_x, bp_y):\n",
        "            if nt_x <= len(mRNA) and nt_y > len(mRNA): #nt_x is mRNA, nt_y is rRNA, they are bound.\n",
        "                most_5p_mRNA = nt_x #starts counting from 0\n",
        "                break\n",
        "\n",
        "        #Extract the base pairings that are 3' of the most_5p_mRNA base pairing\n",
        "        bp_x_3p = []\n",
        "        bp_y_3p = []\n",
        "        for (nt_x, nt_y) in zip(bp_x, bp_y):\n",
        "            if nt_x >= most_5p_mRNA:\n",
        "                bp_x_3p.append(nt_x)\n",
        "                bp_y_3p.append(nt_y)\n",
        "\n",
        "        #Create the mRNA subsequence\n",
        "        mRNA_subsequence = mRNA[0:max(0,most_5p_mRNA - self.standby_site_length - 1)]\n",
        "        standby_site = mRNA[most_5p_mRNA - self.standby_site_length - 1:most_5p_mRNA]\n",
        "\n",
        "        #Fold it and extract the base pairings\n",
        "        if (len(mRNA_subsequence)) > 0:\n",
        "            fold = NuPACK([mRNA_subsequence], material = self.RNA_model)\n",
        "            fold.mfe([1], dangles = self.dangles, Temp = self.temp)\n",
        "            energy_after_5p = fold[\"mfe_energy\"][0]\n",
        "            bp_x_5p = fold[\"mfe_basepairing_x\"][0]   #[0] added 12/13/07\n",
        "            bp_y_5p = fold[\"mfe_basepairing_y\"][0]\n",
        "        else:\n",
        "             bp_x_5p = []\n",
        "             bp_y_5p = []\n",
        "             energy_after_5p = 0.0\n",
        "\n",
        "        #Put the sets of base pairings together\n",
        "        bp_x_after = []\n",
        "        bp_y_after = []\n",
        "        for (nt_x, nt_y) in zip(bp_x_5p, bp_y_5p):\n",
        "            bp_x_after.append(nt_x)\n",
        "            bp_y_after.append(nt_y)\n",
        "\n",
        "        for (nt_x, nt_y) in zip(bp_x_3p, bp_y_3p):\n",
        "            bp_x_after.append(nt_x)\n",
        "            bp_y_after.append(nt_y)\n",
        "\n",
        "        #Calculate its energy\n",
        "        fold = NuPACK([mRNA, self.rRNA], material = self.RNA_model)\n",
        "        energy_after = fold.energy([1, 2], bp_x_after, bp_y_after, dangles = self.dangles, Temp = self.temp)\n",
        "\n",
        "        dG_standby_site = energy_before - energy_after\n",
        "\n",
        "        if (dG_standby_site > 0.0): dG_standby_site = 0.0\n",
        "\n",
        "        index = structure[\"MinStructureID\"]\n",
        "        structure[\"bp_x\"] = bp_x_after\n",
        "        structure[\"bp_y\"] = bp_y_after\n",
        "        structure[\"subopt_basepairing_x\"][index] = bp_x_after\n",
        "        structure[\"subopt_basepairing_y\"][index] = bp_y_after\n",
        "        structure[\"subopt_energy\"][index] = energy_after\n",
        "        structure[\"dG_mRNA_rRNA_corrected\"] = energy_after\n",
        "\n",
        "        return (dG_standby_site, structure)\n",
        "\n",
        "    def calc_dG_mRNA(self,start_pos):\n",
        "        \"\"\"Calculates the dG_mRNA given the mRNA sequence.\"\"\"\n",
        "\n",
        "        mRNA = self.mRNA_input[max(0,start_pos-self.cutoff):min(len(self.mRNA_input),start_pos+self.cutoff)]\n",
        "        fold = NuPACK([mRNA],self.RNA_model)\n",
        "        fold.mfe([1], Temp = self.temp, dangles = self.dangles)\n",
        "\n",
        "        structure = fold\n",
        "        structure[\"mRNA\"] = mRNA\n",
        "        structure[\"bp_x\"] = fold[\"mfe_basepairing_x\"][0]\n",
        "        structure[\"bp_y\"] = fold[\"mfe_basepairing_y\"][0]\n",
        "        structure[\"dG_mRNA\"] = fold[\"mfe_energy\"][0]\n",
        "        structure[\"MinStructureID\"] = 0\n",
        "\n",
        "        dG_mRNA_folding = fold[\"mfe_energy\"][0]\n",
        "\n",
        "        return (dG_mRNA_folding, structure)\n",
        "\n",
        "    def calc_dG_rRNA(self):\n",
        "        \"\"\"Calculates the dG of folding for the last 9 nt of the 16S rRNA. Not used in the free energy model.\"\"\"\n",
        "        fold = NuPACK([self.rRNA],self.RNA_model)\n",
        "        fold.mfe([1], Temp = self.temp, dangles = \"all\")\n",
        "        dG_rRNA_folding = fold[\"mfe_energy\"][0]\n",
        "        return dG_rRNA_folding\n",
        "\n",
        "    def calc_dG_SDopen(self, mRNA_structure, mRNA_rRNA_structure):\n",
        "        \"\"\"Calculate the dG required to unfold the nucleotides in the 16S rRNA binding site.\"\"\"\n",
        "\n",
        "        mRNA = mRNA_structure[\"mRNA\"]\n",
        "        program = mRNA_structure[\"program\"]\n",
        "        index = mRNA_structure[\"MinStructureID\"]\n",
        "        dG_mRNA = mRNA_structure[program + \"_energy\"][index]\n",
        "\n",
        "        index = mRNA_rRNA_structure[\"MinStructureID\"]\n",
        "        bp_x_1 = mRNA_rRNA_structure[\"subopt_basepairing_x\"][index][:]\n",
        "        bp_y_1 = mRNA_rRNA_structure[\"subopt_basepairing_y\"][index][:]\n",
        "\n",
        "        most_5p_mRNA = self.infinity\n",
        "        most_3p_mRNA = -self.infinity\n",
        "\n",
        "        for (nt_x, nt_y) in zip(bp_x_1, bp_y_1):\n",
        "            if nt_y > len(mRNA): #nt is rRNA\n",
        "                most_5p_mRNA = min(most_5p_mRNA, bp_x_1[bp_y_1.index(nt_y)])\n",
        "                most_3p_mRNA = max(most_3p_mRNA, bp_x_1[bp_y_1.index(nt_y)])\n",
        "\n",
        "        pre_mRNA = mRNA[0:most_5p_mRNA]\n",
        "        post_mRNA = mRNA[most_3p_mRNA+1:len(mRNA)+1]\n",
        "\n",
        "        pre_fold = NuPACK([pre_mRNA],material = self.RNA_model)\n",
        "        pre_fold.mfe([1],dangles = self.dangles, Temp = self.temp)\n",
        "        dG_pre = pre_fold[\"mfe_energy\"][0]\n",
        "\n",
        "        post_fold = NuPACK([post_mRNA],material = self.RNA_model)\n",
        "        post_fold.mfe([1],dangles = self.dangles, Temp = self.temp)\n",
        "        dG_post = post_fold[\"mfe_energy\"][0]\n",
        "\n",
        "        energy = dG_pre + dG_post\n",
        "\n",
        "        ddG_mRNA = energy - dG_mRNA #positive if work is required to unfold SD sequence\n",
        "        return ddG_mRNA\n",
        "\n",
        "    def calc_kinetic_score(self, structure = None, mRNA_in = None, bp_x_in = None, bp_y_in = None):\n",
        "        \"\"\"Calculate a \"kinetic score\", a heuristic measure of the maximum time required for the mRNA secondary structure to form. This is related to the RNA polymer model by David et. al. This heuristic should not be used in any way to quantify the folding kinetics of an mRNA sequence because it completely ignores cooperative RNA folding mechanisms, such as zipping or strand displacement. Here, we use it to eliminate mRNA sequences that MAY fold slowly.\"\"\"\n",
        "\n",
        "        if not (structure is None):\n",
        "            program = structure[\"program\"]\n",
        "            mRNA = structure[\"mRNA\"]\n",
        "            index = structure[\"MinStructureID\"]\n",
        "            bp_x = structure[program + \"_basepairing_x\"][index]\n",
        "            bp_y = structure[program + \"_basepairing_y\"][index]\n",
        "\n",
        "        if not (bp_x_in is None) and not (bp_y_in is None) and not (mRNA_in is None):\n",
        "            mRNA = mRNA_in[:]\n",
        "            bp_x = bp_x_in[:]\n",
        "            bp_y = bp_y_in[:]\n",
        "\n",
        "        largest_range_helix = 0\n",
        "        for (nt_x, nt_y) in zip(bp_x, bp_y):\n",
        "            if nt_x <= len(mRNA) and nt_y <= len(mRNA):\n",
        "                val = nt_y - nt_x\n",
        "                largest_range_helix = max(val, largest_range_helix)\n",
        "\n",
        "        kinetic_score = float(largest_range_helix) / float(len(mRNA))\n",
        "        if float(largest_range_helix) > 0:\n",
        "            min_bp_prob = float(largest_range_helix)**(-1.44) #From David et. al.\n",
        "        else:\n",
        "            min_bp_prob = 1.0\n",
        "\n",
        "        return (kinetic_score, min_bp_prob)\n",
        "\n",
        "    def calc_most_5p_mRNA(self,structure_old):\n",
        "        \"\"\"Calculates the most 5' nucleotide in the 16S rRNA binding site.\"\"\"\n",
        "\n",
        "        import copy\n",
        "        structure = copy.deepcopy(structure_old)\n",
        "        mRNA = structure[\"mRNA\"]\n",
        "        bp_x = structure[\"bp_x\"]\n",
        "        bp_y = structure[\"bp_y\"]\n",
        "\n",
        "        #Identify the most 5p mRNA nt that is bound to rRNA\n",
        "        for (nt_x, nt_y) in zip(bp_x, bp_y):\n",
        "            if nt_x <= len(mRNA) and nt_y > len(mRNA): #nt_x is mRNA, nt_y is rRNA, they are bound.\n",
        "                most_5p_mRNA = nt_x\n",
        "                break\n",
        "\n",
        "        return most_5p_mRNA\n",
        "\n",
        "    def calc_longest_helix(self,structure):\n",
        "        \"\"\"Calculate the longest helical structure (longest contiguous list of base pairings) in the secondary structure\"\"\"\n",
        "\n",
        "        mRNA = structure[\"mRNA\"]\n",
        "        bp_x = structure[\"bp_x\"]\n",
        "        bp_y = structure[\"bp_y\"]\n",
        "\n",
        "        longest_helix = 0\n",
        "        helix_length = 1\n",
        "\n",
        "        for (nt_x, nt_y) in zip(bp_x, bp_y):\n",
        "            if (bp_x.count(nt_x+1) > 0 and bp_y.count(nt_y-1) > 0):\n",
        "                helix_length += 1\n",
        "            else:\n",
        "                longest_helix = max(longest_helix, helix_length)\n",
        "                helix_length = 1\n",
        "\n",
        "        return longest_helix\n",
        "\n",
        "    def calc_longest_loop_bulge(self,structure,output_start_end=False,InRBSOnly=False,RBS=None):\n",
        "        \"\"\"Calculate the longest helical loop and bulge structure (longest contiguous list of un-base paired nucleotides starting and ending with a helix (loop -> same helix, bulge -> different helix) in the secondary structure\"\"\"\n",
        "\n",
        "        mRNA = structure[\"mRNA\"]\n",
        "        bp_x = structure[\"bp_x\"]\n",
        "        bp_y = structure[\"bp_y\"]\n",
        "\n",
        "        loop_length = 0\n",
        "        begin_helix = 1\n",
        "\n",
        "        bulge_loop_list = []\n",
        "        helical_loop_list = []\n",
        "\n",
        "        if output_start_end:\n",
        "            bulge_loop_start_end = []\n",
        "            helical_loop_start_end = []\n",
        "\n",
        "        if InRBSOnly and RBS is not None:\n",
        "            RBS_begin = mRNA.find(RBS)\n",
        "            RBS_end = RBS_begin + len(RBS)\n",
        "            nucleotide_range = range(RBS_begin,RBS_end+1)\n",
        "\n",
        "        else:\n",
        "            nucleotide_range = range(1,len(mRNA)+1)\n",
        "\n",
        "        #Find loops. Find bulges.\n",
        "        for n in nucleotide_range:\n",
        "            if bp_x.count(n) == 0 and bp_y.count(n) == 0: #nth nucleotide is not base-paired.\n",
        "\n",
        "                #Determine if nearest neighbor nucleotides are base-paired\n",
        "                (x1,x2,y1,y2) = (bp_x.count(n-1), bp_x.count(n+1), bp_y.count(n-1), bp_y.count(n+1))\n",
        "\n",
        "                #print \"#\", n, (x1,x2,y1,y2)\n",
        "\n",
        "                #middle unpaired nt\n",
        "                if (x1,x2,y1,y2) == (0,0,0,0):\n",
        "                    loop_length += 1\n",
        "\n",
        "                #single mismatch -- loop\n",
        "                elif (x1,x2,y1,y2) == (1,0,0,1) or (x1,x2,y1,y2) == (0,1,1,0):\n",
        "                    loop_length += 1\n",
        "                    begin_helix = n - 1\n",
        "                    end_helix = n + 1\n",
        "\n",
        "                #single mismatch -- bulge\n",
        "                elif (x1,x2,y1,y2) == (1,1,0,0) or (x1,x2,y1,y2) == (0,0,1,1):\n",
        "                    loop_length += 1\n",
        "                    begin_helix = n - 1\n",
        "                    end_helix = n + 1\n",
        "\n",
        "                #starting unpaired nt\n",
        "                elif (x1,x2,y1,y2) == (1,0,0,0) or (x1,x2,y1,y2) == (0,0,1,0):\n",
        "                    loop_length += 1\n",
        "                    begin_helix = n - 1\n",
        "\n",
        "                #ending unpaired nt\n",
        "                elif (x1,x2,y1,y2) == (0,1,0,0) or (x1,x2,y1,y2) == (0,0,0,1):\n",
        "                    loop_length += 1\n",
        "                    end_helix = n + 1\n",
        "\n",
        "                #1,0,1,0 is impossible w/o psuedoknots\n",
        "                #0,1,0,1 is impossible w/o psuedoknots\n",
        "                #Also, all binary combinations with 3 or 4 true are impossible (n-1 or n+1 can not be in both bp_x and bp_y)\n",
        "\n",
        "\n",
        "            elif loop_length > 0:\n",
        "                #Bulge or loop?\n",
        "                #loop\n",
        "\n",
        "                #print \"begin = \", begin_helix\n",
        "                #print \"end = \", end_helix\n",
        "\n",
        "                if ( bp_x.count(begin_helix) > 0 and bp_y.count(end_helix) > 0 and bp_x.index(begin_helix) == bp_y.index(end_helix) ):\n",
        "                    helical_loop_list.append(loop_length)\n",
        "                    loop_length = 0\n",
        "\n",
        "                    if output_start_end:\n",
        "                        #Also return the starting and ending positions of each loop/bulge\n",
        "                        helical_loop_start_end.append((begin_helix,end_helix))\n",
        "\n",
        "                else:\n",
        "\n",
        "                    bp_end = 0\n",
        "                    bp_begin = 0\n",
        "\n",
        "                    if (bp_x.count(end_helix) > 0): bp_begin = bp_y[bp_x.index(end_helix)]\n",
        "                    if (bp_y.count(end_helix) > 0): bp_end = bp_x[bp_y.index(end_helix)]\n",
        "\n",
        "                    if (bp_x.count(begin_helix) > 0): bp_end = bp_y[bp_x.index(begin_helix)]\n",
        "                    if (bp_y.count(begin_helix) > 0): bp_begin = bp_x[bp_y.index(begin_helix)]\n",
        "\n",
        "                    if bp_end > bp_begin:\n",
        "                        bulge_loop_list.append(loop_length)\n",
        "                        loop_length = 0\n",
        "\n",
        "                        if output_start_end:\n",
        "                            #Also return the starting and ending positions of each loop/bulge\n",
        "                            bulge_loop_start_end.append((begin_helix,end_helix))\n",
        "\n",
        "                    else:\n",
        "                        loop_length = 0\n",
        "        if output_start_end:\n",
        "            return (helical_loop_list, bulge_loop_list, helical_loop_start_end, bulge_loop_start_end)\n",
        "        else:\n",
        "            return (helical_loop_list, bulge_loop_list)\n",
        "\n",
        "    def cpu_time(self):\n",
        "        import resource\n",
        "        return resource.getrusage(resource.RUSAGE_SELF)[0]\n",
        "\n",
        "    def combsort(self, input_list):\n",
        "        \"\"\"Sorts the input_list in increasing order (minimum first) according to the comb sort algorithm from Wikipedia. Outputs the corresponding sorted index. \"\"\"\n",
        "\n",
        "        index = range(len(input_list))\n",
        "\n",
        "        #Implementation of comb sort (from Wikipedia)\n",
        "        shrink_factor = 1.24733095\n",
        "\n",
        "        #Init\n",
        "        gap = len(input_list)\n",
        "        swaps = False\n",
        "\n",
        "        while not (gap <= 1 and not swaps):\n",
        "\n",
        "            #update the gap value for a next comb\n",
        "            if gap > 1:\n",
        "                gap = int(gap / shrink_factor)\n",
        "                if gap == 10 or gap == 9:\n",
        "                    gap = 11\n",
        "\n",
        "            i = 0\n",
        "            swaps = False #see bubblesort for an explanation\n",
        "\n",
        "            #a single \"comb\" over the input list\n",
        "            while gap + i < len(input_list):\n",
        "                val = input_list[i]\n",
        "                if val > input_list[i+gap]:\n",
        "                    #Swap values -- verify that lists are being altered correctly\n",
        "                    input_list[i] = input_list[gap + i]\n",
        "                    input_list[gap + i] = val\n",
        "\n",
        "                    #Swap indices of index list accordingly\n",
        "                    temp = index[i]\n",
        "                    index[i] = index[gap + i]\n",
        "                    index[gap + i] = temp\n",
        "\n",
        "                    swaps = True\n",
        "\n",
        "                i += 1\n",
        "\n",
        "        return index\n",
        "\n",
        "    def calc_dG(self):\n",
        "        \"\"\"Calculates each dG term in the free energy model and sums them together to create dG_total\"\"\"\n",
        "\n",
        "        start = self.cpu_time()\n",
        "\n",
        "        #Initialization of data structures\n",
        "        self.start_pos_list = []\n",
        "        self.dG_total_list = []\n",
        "        self.dG_mRNA_list = []\n",
        "        self.dG_mRNA_rRNA_list = []\n",
        "        self.fold_x_list = []\n",
        "        self.fold_y_list = []\n",
        "        self.dG_start_energy_list = []\n",
        "        self.dG_spacing_list = []\n",
        "        self.mRNA_structure_list = []\n",
        "        self.mRNA_rRNA_uncorrected_structure_list = []\n",
        "        self.mRNA_rRNA_corrected_structure_list = []\n",
        "        self.dG_standby_site_list = []\n",
        "        self.dG_spacer_site_list = []\n",
        "        self.kinetic_score_list = []\n",
        "        self.min_bp_prob_list = []\n",
        "        self.longest_helix_list = []\n",
        "        self.three_state_indicator_list = []\n",
        "        self.helical_loop_list_list = []\n",
        "        self.bulge_loop_list_list = []\n",
        "\n",
        "        self.dS1_list = []\n",
        "        self.dS2_list = []\n",
        "        self.most_5p_mRNA_list = []\n",
        "        self.Expression_list = []\n",
        "\n",
        "        for (start_pos, codon) in self.find_start_codons(self.mRNA_input):\n",
        "\n",
        "            try:\n",
        "\n",
        "                #print \"Top of calc_dG here\"\n",
        "\n",
        "                #Set dangles based on length between 5' end of mRNA and start codon\n",
        "                if self.auto_dangles:\n",
        "\n",
        "                    if start_pos > self.cutoff:\n",
        "                        self.dangles = \"none\"\n",
        "\n",
        "                    else:\n",
        "                        self.dangles = \"all\"\n",
        "\n",
        "                else:\n",
        "                    self.dangles = self.dangles_default\n",
        "                    #print \"Auto Dangles set to \", self.dangles\n",
        "\n",
        "                #Start codon energy\n",
        "                dG_start_codon = self.start_codon_energies[codon]\n",
        "\n",
        "                #Energy of mRNA folding\n",
        "                [dG_mRNA,mRNA_structure] = self.calc_dG_mRNA(start_pos)\n",
        "\n",
        "                #Energy of mRNA:rRNA hybridization & folding\n",
        "                [dG_mRNA_rRNA_withspacing,mRNA_rRNA_structure] = self.calc_dG_mRNA_rRNA(start_pos)\n",
        "\n",
        "                dG_mRNA_rRNA_nospacing = mRNA_rRNA_structure[\"dG_mRNA_rRNA\"]\n",
        "\n",
        "                #Standby site correction:\n",
        "                [dG_standby_site, corrected_structure] = self.calc_dG_standby_site(mRNA_rRNA_structure, rRNA_binding = True)\n",
        "\n",
        "                #Total energy is mRNA:rRNA + start - rRNA - mRNA - standby_site\n",
        "                dG_total = dG_mRNA_rRNA_withspacing + dG_start_codon - dG_mRNA - dG_standby_site\n",
        "\n",
        "                #Calculate 'kinetic score': directly related to probability of base pair formation\n",
        "                (kinetic_score,min_bp_prob) = self.calc_kinetic_score(mRNA_structure)\n",
        "\n",
        "                #Calculate dG to open SD sequence\n",
        "                ddG_SD_open = self.calc_dG_SDopen(mRNA_structure, mRNA_rRNA_structure)\n",
        "\n",
        "\n",
        "                self.mRNA_structure_list.append(mRNA_structure)\n",
        "                self.mRNA_rRNA_uncorrected_structure_list.append(mRNA_rRNA_structure)\n",
        "                self.mRNA_rRNA_corrected_structure_list.append(corrected_structure)\n",
        "                self.most_5p_mRNA_list.append(self.calc_most_5p_mRNA(mRNA_rRNA_structure))\n",
        "\n",
        "                (helical_loop_list, bulge_loop_list) = self.calc_longest_loop_bulge(structure=mRNA_structure)\n",
        "                self.longest_helix_list.append(self.calc_longest_helix(structure=mRNA_structure))\n",
        "\n",
        "                self.dG_start_energy_list.append(dG_start_codon)\n",
        "                self.dG_mRNA_list.append(dG_mRNA)\n",
        "                self.dG_mRNA_rRNA_list.append(dG_mRNA_rRNA_nospacing)\n",
        "                self.dG_spacing_list.append(mRNA_rRNA_structure[\"dG_spacing\"])\n",
        "                self.dG_standby_site_list.append(dG_standby_site)\n",
        "                self.dG_total_list.append(dG_total)\n",
        "\n",
        "                self.helical_loop_list_list.append(helical_loop_list)\n",
        "                self.bulge_loop_list_list.append(bulge_loop_list)\n",
        "\n",
        "                self.min_bp_prob_list.append(min_bp_prob)\n",
        "                self.kinetic_score_list.append(kinetic_score)\n",
        "                self.three_state_indicator_list.append(ddG_SD_open)\n",
        "\n",
        "                #Start positions\n",
        "                self.start_pos_list.append(start_pos)\n",
        "\n",
        "                #Expression levels\n",
        "                self.Expression_list.append(self.calc_expression_level(dG_total))\n",
        "\n",
        "                #For exporting the relevant structure to a PDF\n",
        "                #index = mRNA_rRNA_structure[\"MinStructureID\"]\n",
        "                #mRNA_rRNA_structure.export_PDF(index, name = self.name + \": Before standby site\", filename =  self.name + \"_Before_Standby_rRNA.pdf\", program = \"subopt\")\n",
        "\n",
        "                #index = corrected_structure[\"MinStructureID\"]\n",
        "                #corrected_structure.export_PDF(index, name = self.name + \": After standby site\", filename =  self.name + \"_After_Standby_rRNA.pdf\", program = \"subopt\")\n",
        "\n",
        "            except CalcError, msg:\n",
        "                print msg\n",
        "                self.mRNA_structure_list.append([])\n",
        "                self.mRNA_rRNA_uncorrected_structure_list.append([])\n",
        "                self.mRNA_rRNA_corrected_structure_list.append([])\n",
        "\n",
        "                self.most_5p_mRNA_list.append(self.infinity)\n",
        "                self.longest_helix_list.append(self.infinity)\n",
        "\n",
        "                self.dG_start_energy_list.append(self.infinity)\n",
        "                self.dG_mRNA_list.append(self.infinity)\n",
        "                self.dG_mRNA_rRNA_list.append(self.infinity)\n",
        "                self.dG_spacing_list.append(self.infinity)\n",
        "                self.dG_standby_site_list.append(self.infinity)\n",
        "                self.dG_total_list.append(self.infinity)\n",
        "\n",
        "                self.helical_loop_list_list.append(self.infinity)\n",
        "                self.bulge_loop_list_list.append(self.infinity)\n",
        "\n",
        "                self.kinetic_score_list.append(self.infinity)\n",
        "                self.three_state_indicator_list.append(self.infinity)\n",
        "\n",
        "        self.run = 1\n",
        "\n",
        "        end = self.cpu_time()\n",
        "        self.run_time = end - start\n",
        "\n",
        "    def calc_expression_level(self, dG):\n",
        "\n",
        "        import math\n",
        "        return RBS_Calculator.K * math.exp(-dG / RBS_Calculator.RT_eff)\n",
        "\n",
        "    def print_dG(self,max_dG = 1e12,brief = 0, return_string = False, print_expression = False):\n",
        "        '''Print out useful information about the mRNA sequence'''\n",
        "\n",
        "        import math\n",
        "\n",
        "        print_string = \"\"\n",
        "        if self.run == 1:\n",
        "\n",
        "            print \"mRNA Sequence: \", self.name\n",
        "            if return_string: print_string = print_string + \"mRNA Sequence: \" + self.name + \"\\n\"\n",
        "\n",
        "            if len(self.start_position_list) == 0:\n",
        "                print \"No start codons found in input mRNA sequence\"\n",
        "                print \"----------------------------------------------------------------------------------------\"\n",
        "                if return_string: print_string = print_string + \"No start codons found in input mRNA sequence\" + \"\\n\" + \"----------------------------------------------------------------------------------------\" + \"\\n\"\n",
        "            elif len( [dG_total for dG_total in self.dG_total_list if dG_total < max_dG] ) == 0:\n",
        "                print \"No RBSs found with dG <\", str(max_dG), \".\"\n",
        "                print \"----------------------------------------------------------------------------------------\" + \"\\n\"\n",
        "                if return_string: print_string = print_string + \"No RBSs found with dG <\" + str(max_dG) + \".\" + \"\\n\" + \"----------------------------------------------------------------------------------------\" + \"\\n\"\n",
        "\n",
        "            else:\n",
        "\n",
        "                if print_expression:\n",
        "                    Headers = (\"Start\", \"(pos)\", \"Expression level\",\"Kinetic score\")\n",
        "                    format = \"%4s %4s %15s %15s\"\n",
        "                else:\n",
        "                    Headers = (\"Start\",\"(pos)\", \"dG total\", \"dG (rRNA:mRNA)\", \"dG (mRNA)\", \"dG (spacing)\", \"dG (standby)\", \"Kinetic Score\")\n",
        "                    format = \"%4s %4s %12s %12s %12s %12s %12s %12s\"\n",
        "\n",
        "                print format % Headers\n",
        "\n",
        "                if return_string: print_string = print_string + format % Headers + \"\\n\"\n",
        "\n",
        "                for (start,counter) in zip(self.start_position_list,range(len(self.start_position_list))):\n",
        "\n",
        "                    dG_total = self.dG_total_list[counter]\n",
        "                    Expression = RBS_Calculator.K * math.exp(-dG_total / RBS_Calculator.RT_eff)\n",
        "\n",
        "                    if (dG_total < max_dG):\n",
        "\n",
        "                        if (dG_total > self.infinity):\n",
        "                            dG_total = \"Inf\"\n",
        "                            Expression = 0.0\n",
        "\n",
        "                        dG_mRNA = self.dG_mRNA_list[counter]\n",
        "                        if (dG_mRNA > self.infinity): dG_mRNA = \"Inf\"\n",
        "\n",
        "                        dG_mRNA_rRNA = self.dG_mRNA_rRNA_list[counter]\n",
        "                        if (dG_mRNA_rRNA > self.infinity): dG_mRNA_rRNA = \"Inf\"\n",
        "\n",
        "                        dG_spacing = self.dG_spacing_list[counter]\n",
        "                        if (dG_spacing > self.infinity): dG_spacing = \"Inf\"\n",
        "\n",
        "                        dG_standby_site = self.dG_standby_site_list[counter]\n",
        "                        if (dG_standby_site > self.infinity): dG_standby_site = \"Inf\"\n",
        "\n",
        "                        start_codon = self.start_codon_list[counter]\n",
        "\n",
        "                        dG_rRNA = self.dG_rRNA\n",
        "\n",
        "                        kinetic_score = self.kinetic_score_list[counter]\n",
        "\n",
        "                        if print_expression:\n",
        "                            print format % (start_codon, str(start), str(round(Expression,2)), str(round(kinetic_score,2)))\n",
        "\n",
        "                        else:\n",
        "                            print format % (start_codon, str(start), str(dG_total), str(dG_mRNA_rRNA), str(dG_mRNA), str(dG_spacing), str(dG_standby_site), str(round(kinetic_score,2)))\n",
        "\n",
        "                        if return_string: print_string = print_string + format % (start_codon, str(start), str(dG_total), str(dG_mRNA_rRNA), str(dG_mRNA), str(dG_spacing), str(dG_standby_site), str(round(kinetic_score,2))) + \"\\n\"\n",
        "\n",
        "                print \"----------------------------------------------------------------------------------------\"\n",
        "                if return_string: print_string = print_string + \"----------------------------------------------------------------------------------------\" + \"\\n\"\n",
        "\n",
        "                #print \"Computation Time: \", str(self.run_time), \" seconds.\"\n",
        "\n",
        "                if return_string: return print_string\n",
        "        else:\n",
        "            raise RuntimeError(\"The RBS Calculator has not been run yet. Call the 'calc_dG' method.\")\n",
        "\n",
        "    def save_data(self, handle, header = False):\n",
        "\n",
        "        infinity = 1e6\n",
        "\n",
        "        if self.run == 1:\n",
        "\n",
        "            if (header):\n",
        "                #Print header information --> all the parameters\n",
        "                parameters = \"RNA model: \\t%s \\nDangles: \\t%s \\nTemperature: \\t%s \\nOptimal Spacing \\t%s \\nSpacing constant (push) \\t%s\\nSpacing constant (pull) \\t%s\\nCutoff \\t%s\\nStandby Site Length \\t%s\\nEnergy Cutoff \\t%s\\nrRNA sequence \\t%s\\n\" % (self.RNA_model, str(self.dangles), str(self.temp), str(self.optimal_spacing), str(self.dG_spacing_constant_push), str(self.dG_spacing_constant_pull), str(self.cutoff), str(self.standby_site_length), str(self.energy_cutoff), self.rRNA)\n",
        "\n",
        "                handle.writelines(parameters)\n",
        "\n",
        "            #Name, dG total, dG rRNA:mRNA, dG mRNA, dG spacing, dG standby, dG start codon, kinetic score, longest helix, longest loop, start position\n",
        "            #format = \"%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\n\"\n",
        "            format = \"%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\t%10s\\n\"\n",
        "\n",
        "            for counter in range(len(self.start_position_list)):\n",
        "\n",
        "                dG_total = self.dG_total_list[counter]\n",
        "                if (dG_total >= infinity): dG_total = \"Inf\"\n",
        "\n",
        "                dG_mRNA = self.dG_mRNA_list[counter]\n",
        "                if (dG_mRNA >= infinity): dG_mRNA = \"Inf\"\n",
        "\n",
        "                dG_mRNA_rRNA = self.dG_mRNA_rRNA_list[counter]\n",
        "                if (dG_mRNA_rRNA >= infinity): dG_mRNA_rRNA = \"Inf\"\n",
        "\n",
        "                dG_spacing = self.dG_spacing_list[counter]\n",
        "                if (dG_spacing >= infinity): dG_spacing = \"Inf\"\n",
        "\n",
        "                dG_standby_site = self.dG_standby_site_list[counter]\n",
        "                if (dG_standby_site >= infinity): dG_standby_site = \"Inf\"\n",
        "\n",
        "                kinetic_score = self.kinetic_score_list[counter]\n",
        "\n",
        "                longest_helix = self.longest_helix_list[counter]\n",
        "                if len(self.helical_loop_list_list[counter])>0:\n",
        "                    longest_loop = max(self.helical_loop_list_list[counter])\n",
        "                else:\n",
        "                    longest_loop = 0\n",
        "\n",
        "                most_5p_mRNA = self.most_5p_mRNA_list[counter]\n",
        "\n",
        "                three_state = self.three_state_indicator_list[counter]\n",
        "\n",
        "                handle.writelines(format % (self.name.split(\" \")[0],dG_total, dG_mRNA_rRNA,dG_mRNA,dG_spacing,dG_standby_site, self.start_codon_energies[self.start_codon_list[counter]],round(kinetic_score,2), str(longest_helix), str(self.start_position_list[counter]), str(most_5p_mRNA), str(three_state) ))\n",
        "\n",
        "        else:\n",
        "            raise RuntimeError(\"The RBS Calculator has not been run yet. Call the 'calc_dG' method.\")\n",
        "\n",
        "#----------------------------------------------------------------------------------------------------------\n",
        "#End RBS_Calculator class\n",
        "#----------------------------------------------------------------------------------------------------------\n",
        "def calc_dG_from_file(handle, output, verbose = True, parameters = {}):\n",
        "    from Bio import SeqIO\n",
        "    from RBS_Calculator import RBS_Calculator\n",
        "\n",
        "    records = SeqIO.parse(handle,\"fasta\")\n",
        "    First = True\n",
        "    export_PDF = True\n",
        "\n",
        "    #for i in range(30):\n",
        "    #    records.next()\n",
        "\n",
        "    for record in records:\n",
        "\n",
        "        mRNA = record.seq.tostring().upper()\n",
        "\n",
        "        #Set any defaults\n",
        "        start_range = [0, len(mRNA)]\n",
        "        name = record.description.split(\" \")[0]\n",
        "\n",
        "        #Create instance of RBS Calculator\n",
        "        test = RBS_Calculator(mRNA, start_range, name)\n",
        "\n",
        "        #Examine kvars dictionary and pull out any options. Assign them to instanced class.\n",
        "        for (key,value) in parameters.items():\n",
        "\n",
        "            if key == \"cutoff\":\n",
        "                test.cutoff = value\n",
        "            elif key == \"start_range\":\n",
        "                test.start_range = value\n",
        "            elif key == \"rRNA\":\n",
        "                test.rRNA = value\n",
        "            elif key == \"energy_cutoff\":\n",
        "                test.energy_cutoff = value\n",
        "            elif key == \"standby_site_length\":\n",
        "                test.standby_site_length = value\n",
        "            elif key == \"dangles\":\n",
        "                test.dangles = value\n",
        "            elif key == \"export_PDF\":\n",
        "                export_PDF = value\n",
        "\n",
        "        test.calc_dG()\n",
        "        test.print_dG(test.infinity,print_expression=verbose)\n",
        "        test.save_data(output, First)\n",
        "\n",
        "        if First:\n",
        "            First = False\n",
        "\n",
        "        if export_PDF:\n",
        "            num_structs = len(test.mRNA_rRNA_uncorrected_structure_list)\n",
        "            for (structure,counter) in zip(test.mRNA_rRNA_uncorrected_structure_list,range(num_structs)):\n",
        "                index = structure[\"MinStructureID\"]\n",
        "                structure.export_PDF(index, name, filename =  name + \"_rRNA\" + \"_\" + str(counter) + \".pdf\", program = \"subopt\")\n",
        "\n",
        "            num_structs = len(test.mRNA_structure_list)\n",
        "            for (structure,counter) in zip(test.mRNA_structure_list,range(num_structs)):\n",
        "                structure.export_PDF(0, name, filename = name + \"_mRNA\" + \"_\" + str(counter) + \".pdf\")\n",
        "\n",
        "    output.close()\n",
        "\n",
        "def calc_dG_pre_post_RBS(pre_list,post_list,RBS_list,name_list,output,verbose = True, parameters = {}):\n",
        "\n",
        "    from RBS_Calculator import RBS_Calculator\n",
        "\n",
        "    First = True\n",
        "\n",
        "    for (pre,post,RBS,name) in zip(pre_list,post_list,RBS_list,name_list):\n",
        "\n",
        "        mRNA = pre + RBS + post\n",
        "\n",
        "        start_range = [0, len(mRNA)]\n",
        "\n",
        "        #Create instance of RBS Calculator\n",
        "        test = RBS_Calculator(mRNA, start_range, name)\n",
        "\n",
        "        #Examine kvars dictionary and pull out any options. Assign them to instanced class.\n",
        "        for (key,value) in parameters.items():\n",
        "\n",
        "            if key == \"cutoff\":\n",
        "                test.cutoff = value\n",
        "            elif key == \"start_range\":\n",
        "                test.start_range = value\n",
        "            elif key == \"rRNA\":\n",
        "                test.rRNA = value\n",
        "            elif key == \"energy_cutoff\":\n",
        "                test.energy_cutoff = value\n",
        "            elif key == \"standby_site_length\":\n",
        "                test.standby_sitRBSe_length = value\n",
        "            elif key == \"dangles\":\n",
        "                test.dangles = value\n",
        "            elif key == \"export_PDF\":\n",
        "                export_PDF = value\n",
        "\n",
        "        test.calc_dG()\n",
        "        if verbose: test.print_dG(test.infinity)\n",
        "\n",
        "        test.save_data(output, First)\n",
        "        if First:\n",
        "            First = False\n",
        "\n",
        "    output.close()\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "\n",
        "    from Bio import SeqIO\n",
        "\n",
        "\n",
        "\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Amin_RBSs.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Forward_Predictions.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Context_Tests_RBSs_2nd.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/DNA20_CDSs.fasta\"\n",
        "    #output_filename = \"/common/RBS_Calculator/Output_Forward_Predictions_dangles_all_TTC.txt\"\n",
        "    #output_filename = \"/common/RBS_Calculator/Output_Context_Tests_RBSs_test.txt\"\n",
        "    filename = \"/common/RBS_Calculator/DataSets/All_Tested_RBSs.txt\"\n",
        "    #filename = \"E:\\RBS_Calculator\\DataSets\\All_Tested_RBSs.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Testing_RBSs.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Karsten_nif_RBSs.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/Dan_RBS_silk_sequences.fasta\"\n",
        "    #output_filename = \"/common/RBS_Calculator/Output_All_Tested_RBSs_newspacing.txt\"\n",
        "    #output_filename = \"/common/RBS_Calculator/Output_DNA20_Designed_RBS_1.txt\"\n",
        "    #output_filename = \"/common/RBS_Calculator/Output_Context_Tests_RBSs_3.txt\"\n",
        "    #output_filename = \"/common/RBS_Calculator/Output_Karsten_nif_RBSs.txt\"\n",
        "    #output_filename = \"/common/RBS_Calculator/Amin_RBSs_Output.txt\"\n",
        "\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Crt_version_RBSs.txt\"\n",
        "\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Output_Crt_natural_RBSs.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/RBS_Spacing.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Ethan_RBSs_all.txt\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Dan_sicP.fasta\"\n",
        "    #filename = \"/common/RBS_Calculator/DataSets/Natural_prg_org_operon.str\"\n",
        "\n",
        "\n",
        "    #output_filename = \"/common/RBS_Calculator/Dan_Output_silk_RBSs.txt\"\n",
        "    #handle = open(filename,\"rU\")\n",
        "    #output = open(output_filename,\"w\")\n",
        "\n",
        "    #verbose = False\n",
        "\n",
        "    #pars = {\"start_range\":[18,71],\"pre_window_length\":0,\"post_window_length\":0,\"export_PDF\":False}\n",
        "    #calc_dG_from_file(handle, output, verbose,pars)\n",
        "    #handle.close()\n",
        "    #output.close()\n",
        "\n",
        "    window_list = range(10,65)\n",
        "    for window in window_list:\n",
        "        print \"Cutoff = \", window\n",
        "        output_filename = \"/common/RBS_Calculator/DataSets/cutoff/Output_All_Tested_RBSs_cutoff_\" + str(window) + \".txt\"\n",
        "\n",
        "        handle = open(filename,\"rU\")\n",
        "        output = open(output_filename,\"w\")\n",
        "\n",
        "        verbose = True\n",
        "\n",
        "        pars = {\"start_range\":[18,44],\"export_PDF\":False,\"cutoff\":window}\n",
        "        calc_dG_from_file(handle, output, verbose,pars)\n",
        "        handle.close()\n",
        "        output.close()\n",
        "\n",
        "    #pre = \"ggggaattgtgagcggataacaattcccctctagaa\"\n",
        "    #RBS = \"GAGGAAGTTAGTAAGGAGGTCAGGCGA\"\n",
        "\n",
        "    #pre_list = []\n",
        "    #RBS_list = []\n",
        "    #CDS_list = []\n",
        "    #name_list = []\n",
        "    ##Get protein coding sequence\n",
        "    #records = SeqIO.parse(handle,\"fasta\")\n",
        "    #for record in records:\n",
        "        #CDS_list.append(record.seq.tostring().upper())\n",
        "        #pre_list.append(pre)\n",
        "        #RBS_list.append(RBS)\n",
        "        #name_list.append(record.description)\n",
        "\n",
        "    #calc_dG_pre_post_RBS(pre_list,CDS_list,RBS_list,name_list,output,verbose = True, parameters = pars)\n",
        "\n",
        "\n",
        "    #name = \"Test\"\n",
        "    #mRNA = \"TTCTAGAGGGGGGATCTCCCCCCAAAAAATAAGAGGTACACATGACTAAAACTTTCAAAGGCTCAGTATTCCCACTGAG\"\n",
        "\n",
        "    #start_range = [0, len(mRNA)]\n",
        "    #test = RBS_Calculator(mRNA, start_range, name)\n",
        "    #test.calc_dG()\n",
        "    #test.print_dG(test.infinity)\n",
        "\n",
        "    ##test.post_window_length = 19\n",
        "    #test.auto_dangles = False\n",
        "    #test.default_dangles = \"all\"\n",
        "    #test.pre_window_length = 1000\n",
        "    #test.post_window_length = 20\n",
        "    #test.cutoff = 50\n",
        "    #test.calc_dG()\n",
        "    #test.print_dG(test.infinity)\n",
        "\n",
        "\n",
        "    #if export_PDF:\n",
        "            #num_structs = len(test.mRNA_rRNA_corrected_structure_list)\n",
        "            #for (structure,counter) in zip(test.mRNA_rRNA_corrected_structure_list,range(num_structs)):\n",
        "\n",
        "                #index = structure[\"MinStructureID\"]\n",
        "                #structure.export_PDF(index, name, filename =  name + \"_rRNA\" + \"_\" + str(counter) + \".pdf\", program = \"energy\")\n",
        "\n",
        "            #for (structure,counter) in zip(test.mRNA_rRNA_uncorrected_structure_list,range(num_structs)):\n",
        "\n",
        "                #index = structure[\"MinStructureID\"]\n",
        "                #structure.export_PDF(index, name, filename =  name + \"_rRNA_uncorrected\" + \"_\" + str(counter) + \".pdf\", program = \"energy\")\n",
        "\n",
        "            #num_structs = len(test.mRNA_structure_list)\n",
        "            #for (structure,counter) in zip(test.mRNA_structure_list,range(num_structs)):\n",
        "                #structure.export_PDF(0, name, filename = name + \"_mRNA\" + \"_\" + str(counter) + \".pdf\")\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "fW9N6ap8kof2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "7cEtHYpVkE_t"
      },
      "outputs": [],
      "source": [
        "#@title code for Run_RBS_Calculator.py\n",
        "#This file is part of the Ribosome Binding Site Calculator.\n",
        "\n",
        "#along with Ribosome Binding Site Calculator.  If not, see <http://www.gnu.org/licenses/>.\n",
        "\n",
        "from RBS_Calculator import RBS_Calculator\n",
        "import sys, math\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    #Read command line arguments\n",
        "    input_args = []\n",
        "    for arg in sys.argv:\n",
        "        input_args.append(arg)\n",
        "\n",
        "    start = 0\n",
        "\n",
        "    if len(input_args) == 3:\n",
        "        seq = input_args[1]\n",
        "        start = input_args[2]\n",
        "    elif len(input_args) == 2:\n",
        "        seq = input_args[1]\n",
        "    else:\n",
        "        output = \"Usage: \" + input_args[0] + \" [RNA/DNA Sequence] (start position)\" + \"\\n\"\n",
        "        print output\n",
        "\n",
        "    if start == 0:\n",
        "        start_range = [0, len(seq)]\n",
        "    else:\n",
        "        start_range = [int(start), int(start)+1]\n",
        "\n",
        "    name = \"no name\"\n",
        "\n",
        "    #Create instance of RBS Calculator\n",
        "    calcObj = RBS_Calculator(seq, start_range, name)\n",
        "    calcObj.calc_dG()\n",
        "\n",
        "    dG_total_list = calcObj.dG_total_list[:]\n",
        "    start_pos_list = calcObj.start_pos_list[:]\n",
        "    kinetic_score_list = calcObj.kinetic_score_list[:]\n",
        "\n",
        "    expr_list = []\n",
        "    for dG in dG_total_list:\n",
        "        expr_list.append(calcObj.K * math.exp(-dG/calcObj.RT_eff))\n",
        "\n",
        "    print len(expr_list)\n",
        "    for (expr,start_pos,ks) in zip(expr_list,start_pos_list,kinetic_score_list):\n",
        "        print start_pos, expr, ks "
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Code RBS_MC_Design\n",
        "\n",
        "#This Python class generates a synthetic RBS sequence according to a user's target translation initiation rate, protein coding sequence, and pre-sequence. RBS constraints are not allowed.\n",
        "#This file is part of the Ribosome Binding Site Calculator.\n",
        "#The Ribosome Binding Site Calculator is free software: you can redistribute it and/or modify\n",
        "\n",
        "\n",
        "from RBS_Calculator import RBS_Calculator\n",
        "import random, math, sets\n",
        "\n",
        "infinity = 1.0e20\n",
        "\n",
        "#Constraints settings\n",
        "max_kinetic_score = 0.50\n",
        "max_three_state_indicator = 6.0\n",
        "min_helical_loop = 4\n",
        "max_helical_loop = 12\n",
        "Max_RBS_Length = RBS_Calculator.cutoff\n",
        "\n",
        "#dG ranges\n",
        "dG_range_high = 25.0\n",
        "dG_range_low = -18.0\n",
        "\n",
        "num_rbs_calculations = 0\n",
        "\n",
        "def dsu_sort(idx, seq):\n",
        "    \"\"\"Sorts a list of tuples according to the idx column using a Decorate-Sort-Undecorate method\"\"\"\n",
        "\n",
        "    for i, e in enumerate(seq):\n",
        "        seq[i] = (e[idx], e)\n",
        "    seq.sort()\n",
        "    seq.reverse()\n",
        "    for i, e in enumerate(seq):\n",
        "        seq[i] = e[1]\n",
        "    return seq\n",
        "\n",
        "def weighted_choice(list_of_tuples):\n",
        "    \"\"\"Randomly chooses from a list of choices according to their weighted probabilities.\"\"\"\n",
        "\n",
        "    n = random.uniform(0.0, 1.0)\n",
        "    for item, weight in list_of_tuples:\n",
        "        if n < weight:\n",
        "            break\n",
        "        n = n - weight\n",
        "    return item\n",
        "\n",
        "def Run_RBS_Calculator(pre_seq,post_seq,RBS,verbose=True):\n",
        "    \"\"\"Short cut function to run the RBS Calculator on a pre_sequence, CDS, and RBS.\"\"\"\n",
        "\n",
        "    if vars().has_key('estimator'): del(estimator)\n",
        "\n",
        "    start_range = [len(pre_seq) + len(RBS) - 2, len(pre_seq) + len(RBS) + 2]\n",
        "\n",
        "    mRNA = pre_seq.upper() + RBS + post_seq.upper()\n",
        "\n",
        "    estimator = RBS_Calculator(mRNA, start_range, \"\")\n",
        "    estimator.calc_dG()\n",
        "    if verbose: estimator.print_dG()\n",
        "\n",
        "    global num_rbs_calculations\n",
        "    num_rbs_calculations+=1\n",
        "\n",
        "    return estimator\n",
        "\n",
        "def Generate_Random_RBS(All_Random = False, Max_length = 20, Pre_length = 5, PchooseSD = 0.5, Core_length = 6, max_nonoptimal_spacing = 5):\n",
        "    \"\"\"Generates a random RBS sequence tailored towards the target translation initiation rate.\"\"\"\n",
        "\n",
        "    RBS = []\n",
        "    if All_Random:\n",
        "\n",
        "        for i in range(Max_length):\n",
        "            RBS.append(random.choice([\"A\",\"T\",\"G\",\"C\"]))\n",
        "        return \"\".join(RBS)\n",
        "\n",
        "    for i in range(Pre_length):\n",
        "        RBS.append(random.choice([\"A\",\"T\",\"G\",\"C\"]))\n",
        "\n",
        "    SD = [\"T\",\"A\",\"A\",\"G\",\"G\",\"A\",\"G\",\"G\",\"T\"]\n",
        "\n",
        "    #Choose Core_length nucleotides. Choose from the SD sequence with probability PchooseSD\n",
        "    #Choose from non-SD sequence with probability (1 - PchooseSD) / 3\n",
        "    #The beginning/end of the Core_length wrt to the SD sequence is uniformly randomly determined\n",
        "\n",
        "    Core_length = min(len(SD),Core_length) #Can't be greater then SD length\n",
        "    diff = len(SD) - Core_length\n",
        "    begin = int(random.random() * diff)\n",
        "\n",
        "    for i in range(Core_length):\n",
        "        rand = random.random()\n",
        "        if rand <= PchooseSD:\n",
        "            RBS.append(SD[begin+i])\n",
        "        else:\n",
        "            choices = [\"A\",\"T\",\"G\",\"C\"]\n",
        "            choices.remove(SD[begin+i])\n",
        "            RBS.append(random.choice(choices))\n",
        "\n",
        "    optimal_spacing = RBS_Calculator.optimal_spacing\n",
        "    offset = diff - begin\n",
        "\n",
        "    spacing = random.choice(range(max(0,offset + optimal_spacing - max_nonoptimal_spacing), offset + optimal_spacing + max_nonoptimal_spacing))\n",
        "\n",
        "    for i in range(spacing):\n",
        "        RBS.append(random.choice([\"A\",\"T\",\"G\",\"C\"]))\n",
        "\n",
        "    if len(RBS) > Max_length:\n",
        "         RBS = RBS[len(RBS)-Max_length:len(RBS)+1]\n",
        "\n",
        "    return \"\".join(RBS)\n",
        "\n",
        "def calc_constraints(RBS,estimator):\n",
        "    \"\"\"Calculates the sequence constraints. Returns True if one is violated. \"\"\"\n",
        "\n",
        "    kinetic_score = estimator.kinetic_score_list[0]\n",
        "    three_state_indicator = estimator.three_state_indicator_list[0]\n",
        "    #(helical_loop_list,bulge_loop_list,helical_start_ends,bulge_start_ends) = estimator.calc_longest_loop_bulge(estimator,True,True,RBS)\n",
        "\n",
        "    #print \"KS = \", kinetic_score\n",
        "    #print \"3-state = \", three_state_indicator\n",
        "    #print \"max/min helical = \", max(helical_loop_list), min(helical_loop_list)\n",
        "\n",
        "    if kinetic_score > max_kinetic_score: return True\n",
        "    if three_state_indicator > max_three_state_indicator: return True\n",
        "    #if min(helical_loop_list) < min_helical_loop: return True\n",
        "    #if max(helical_loop_list) > max_helical_loop: return True\n",
        "\n",
        "    return False\n",
        "\n",
        "def RemoveStartCodons(sequence):\n",
        "    \"\"\"Removes any start codons from an input sequence.\"\"\"\n",
        "\n",
        "    import random\n",
        "    import re\n",
        "\n",
        "    regexp_str = \"|\".join(RBS_Calculator.start_codons)\n",
        "    find_starts = re.compile(regexp_str)\n",
        "    matches = find_starts.finditer(sequence.upper())\n",
        "\n",
        "    new_seq = sequence[:]\n",
        "    for match in matches:\n",
        "        start_pos = match.start()\n",
        "\n",
        "        triplet = []\n",
        "        triplet.append(random.choice(['A','T','G','C']))\n",
        "        triplet.append(random.choice(['A','G','C']))\n",
        "        triplet.append(random.choice(['A','T','C']))\n",
        "\n",
        "        new_seq = new_seq[0:start_pos] + \"\".join(triplet) + new_seq[start_pos+3:len(new_seq)+1]\n",
        "\n",
        "    matches = find_starts.search(new_seq.upper())\n",
        "    if matches is None:\n",
        "        return new_seq\n",
        "    else:\n",
        "        return RemoveStartCodons(new_seq)\n",
        "\n",
        "def compnt(nt):\n",
        "    if (nt.upper() == 'A'): return 'T'\n",
        "    if (nt.upper() == 'T'): return 'A'\n",
        "    if (nt.upper() == 'G'): return 'C'\n",
        "    if (nt.upper() == 'C'): return 'G'\n",
        "\n",
        "def MCmove_lower_kinetic_score(pre_seq,post_seq,RBS,estimator = None,MaxIters=infinity):\n",
        "    \"\"\"Removes long-range base paired nucleotides from an mRNA sequence (pre-seq,CDS,RBS group). Used when generating initial conditions for synthetic RBS sequences.\"\"\"\n",
        "\n",
        "    if estimator is None:\n",
        "        estimator = Run_RBS_Calculator(pre_seq,post_seq,RBS,verbose=False)\n",
        "\n",
        "    counter=0\n",
        "    kinetic_score = estimator.kinetic_score_list[0]\n",
        "    while (counter < MaxIters and kinetic_score > max_kinetic_score):\n",
        "\n",
        "        structure = estimator.mRNA_structure_list[0]\n",
        "        mRNA = structure[\"mRNA\"]\n",
        "        RBS_begin = mRNA.find(RBS)\n",
        "        RBS_end = RBS_begin + len(RBS)\n",
        "\n",
        "        #Alter RBS to reduce kinetic score\n",
        "        #Create a sorted list of bp'd nucleotides with the ones with the highest kinetic score at the top\n",
        "        ks_list = []\n",
        "        bp_x = structure[\"bp_x\"]\n",
        "        bp_y = structure[\"bp_y\"]\n",
        "        for (nt_x,nt_y) in zip(bp_x,bp_y):\n",
        "            ks_list.append( (nt_y - nt_x, nt_x, nt_y) )\n",
        "\n",
        "        dsu_sort(0,ks_list) #Sort max-top according to 1st column: kinetic_score\n",
        "\n",
        "        #Determine the bp'd nucleotides with the highest kinetic score. Are either located in the RBS?\n",
        "        #If so, replace them with another random nucleotide\n",
        "        nucleotides = sets.Set(['A', 'T', 'G', 'C'])\n",
        "\n",
        "        num_mutations = min(len(ks_list),10)\n",
        "        for i in range(num_mutations):\n",
        "            nt_x = ks_list[i][1] - 1 #python index\n",
        "            nt_y = ks_list[i][2] - 1 #python index\n",
        "\n",
        "            #nt_x is in the RBS\n",
        "            if (nt_x >= RBS_begin and nt_x < RBS_end):\n",
        "\n",
        "                pos = nt_x - RBS_begin\n",
        "                letter = random.choice(list(nucleotides ^ sets.Set(RBS[pos])))\n",
        "\n",
        "                #print \"Mutating \", RBS[pos], \" --> \", letter\n",
        "\n",
        "                RBS = RBS[0:pos] + letter + RBS[pos+1:len(RBS)+1]\n",
        "\n",
        "\n",
        "            #nt_y is in the RBS\n",
        "            elif (nt_y >= RBS_begin and nt_y < RBS_end):\n",
        "                pos = nt_y - RBS_begin\n",
        "                letter = random.choice(list(nucleotides ^ sets.Set(RBS[pos])))\n",
        "\n",
        "                #print \"Mutating \", RBS[pos], \" --> \", letter\n",
        "\n",
        "                RBS = RBS[0:pos] + letter + RBS[pos+1:len(RBS)+1]\n",
        "\n",
        "\n",
        "            elif len(RBS) < RBS_Calculator.cutoff:\n",
        "                #Insert a nucleotide at the 5' end of the RBS\n",
        "                letter = random.choice(list(nucleotides))\n",
        "\n",
        "                #print \"Inserting \", letter, \" at 5' end\"\n",
        "\n",
        "                RBS = letter + RBS\n",
        "\n",
        "        RBS = RemoveStartCodons(RBS) #No start codons in RBS!!\n",
        "\n",
        "        #print \"RBS = \", RBS\n",
        "        estimator = Run_RBS_Calculator(pre_seq,post_seq,RBS,verbose=False)\n",
        "        kinetic_score = estimator.kinetic_score_list[0]\n",
        "\n",
        "    return (RBS, estimator)\n",
        "\n",
        "\n",
        "def MCmove_constrain_helical_loop(pre_seq,post_seq,RBS,estimator):\n",
        "    \"\"\"Modifies the mRNA sequence to reduce the size of helical loops.\"\"\"\n",
        "\n",
        "    #Alter RBS sequence so that max/min helical loop constraints are valid\n",
        "\n",
        "    structure = estimator.mRNA_structure_list[0]\n",
        "    (helical_loop_list,bulge_loop_list,helical_start_ends,bulge_start_ends) = estimator.calc_longest_loop_bulge(structure,True,True,RBS)\n",
        "\n",
        "    RBS_begin = len(pre_seq)\n",
        "    RBS_end = RBS_begin + len(RBS)\n",
        "\n",
        "    #Insert or delete nucleotides to increase/decrease loop length\n",
        "    for (loop_length,start_end) in zip(helical_loop_list,helical_start_ends):\n",
        "\n",
        "        if loop_length > max_helical_loop:\n",
        "            #Choose random nucleotide in loop. Delete it.\n",
        "\n",
        "            #Identify what part of the loop is in the RBS\n",
        "            RBS_range = sets.Set(range(RBS_begin+1,RBS_end+1))\n",
        "            loop_range = sets.Set(range(start_end[0]+1,start_end[1]))\n",
        "            change_range = list(RBS_range & loop_range) #Intersection\n",
        "\n",
        "            #print \"Loops in RBS: \", change_range\n",
        "\n",
        "            if len(change_range) > 0:\n",
        "                pos = random.choice(change_range) - len(pre_seq)\n",
        "                RBS = RBS[0:pos] + RBS[pos+1:len(RBS)+1] #Delete nucleotide at position pos\n",
        "\n",
        "\n",
        "        elif loop_length < min_helical_loop:\n",
        "            #Choose random position in loop and insert a nucleotide before it.\n",
        "            #Identify what part of the loop is in the RBS\n",
        "            RBS_range = sets.Set(range(RBS_begin+1,RBS_end+1))\n",
        "            loop_range = sets.Set(range(start_end[0]+1,start_end[1]))\n",
        "            change_range = list(RBS_range & loop_range) #Intersection\n",
        "\n",
        "            #print \"Loops in RBS: \", change_range\n",
        "\n",
        "            if len(change_range) > 0:\n",
        "                pos = random.choice(change_range) - len(pre_seq)\n",
        "                letter = random.choice(['A','T','C','G'])\n",
        "                RBS = RBS[0:pos] + letter + RBS[pos+1:len(RBS)+1]\n",
        "\n",
        "    #estimator = Run_RBS_Calculator(pre_seq,post_seq,RBS,verbose=False)\n",
        "\n",
        "    return RBS\n",
        "\n",
        "def GetInitialRBS(pre_seq,post_seq,dG_target):\n",
        "    \"\"\"Generates a random initial condition for designing a synthetic RBS sequence.\"\"\"\n",
        "\n",
        "    use_new = True\n",
        "    Pre_length = 25\n",
        "\n",
        "    dG_target_nondim = (dG_target - dG_range_high) / (dG_range_low - dG_range_high)\n",
        "    #0.0: Low expression\n",
        "    #1.0: High expression\n",
        "\n",
        "    if     dG_target_nondim <  0.125:\n",
        "        PchooseSD = 0.50\n",
        "        Core_length = 4\n",
        "        max_nonoptimal_spacing = 10\n",
        "    elif   dG_target_nondim <  0.250:\n",
        "        PchooseSD = 0.50\n",
        "        Core_length = 4\n",
        "        max_nonoptimal_spacing = 10\n",
        "    elif   dG_target_nondim < 0.5:\n",
        "        PchooseSD = 0.75\n",
        "        Core_length = 4\n",
        "        max_nonoptimal_spacing = 10\n",
        "    elif   dG_target_nondim < 0.7:\n",
        "        PchooseSD = 0.75\n",
        "        Core_length = 4\n",
        "        max_nonoptimal_spacing = 5\n",
        "    elif   dG_target_nondim <  0.8:\n",
        "        PchooseSD = 0.75\n",
        "        Core_length = 6\n",
        "        max_nonoptimal_spacing = 5\n",
        "    elif   dG_target_nondim <  0.9:\n",
        "        PchooseSD = 0.90\n",
        "        Core_length = 6\n",
        "        max_nonoptimal_spacing = 5\n",
        "    elif   dG_target_nondim <  0.95:\n",
        "        PchooseSD = 0.90\n",
        "        Core_length = 8\n",
        "        max_nonoptimal_spacing = 3\n",
        "    elif   dG_target_nondim <=  1.00:\n",
        "        PchooseSD = 1.0\n",
        "        Core_length = 9\n",
        "        max_nonoptimal_spacing = 2\n",
        "\n",
        "    dG_total = dG_range_high + 1\n",
        "    while dG_total > dG_range_high:\n",
        "        RBS = Generate_Random_RBS(False, Max_RBS_Length, Pre_length, PchooseSD, Core_length, max_nonoptimal_spacing)\n",
        "        RBS = RemoveStartCodons(RBS)\n",
        "\n",
        "        estimator = Run_RBS_Calculator(pre_seq,post_seq,RBS,verbose=False)\n",
        "\n",
        "        if use_new:\n",
        "            RBS = MCmove_constrain_helical_loop(pre_seq,post_seq,RBS,estimator)\n",
        "            (RBS,estimator) = MCmove_lower_kinetic_score(pre_seq,post_seq,RBS,estimator)\n",
        "\n",
        "        else:\n",
        "            counter=0\n",
        "            while dG_total > 0 or calc_constraints(RBS,estimator):\n",
        "                counter+=1\n",
        "                RBS = Generate_Random_RBS(False, Max_RBS_Length, Pre_length, PchooseSD, Core_length, max_nonoptimal_spacing)\n",
        "                RBS = RemoveStartCodons(RBS)\n",
        "\n",
        "                estimator = Run_RBS_Calculator(pre_seq,post_seq,RBS,verbose=False)\n",
        "                dG_total = estimator.dG_total_list[0]\n",
        "\n",
        "        dG_total = estimator.dG_total_list[0]\n",
        "\n",
        "    return (RBS,estimator)\n",
        "\n",
        "def Monte_Carlo_Design(pre_seq, post_seq, RBS_init = None, TIR_target = None, dG_target = None, MaxIter = 10000, verbose = False):\n",
        "    \"\"\"Master function for designing synthetic RBS sequences without constraints.\"\"\"\n",
        "\n",
        "    #Check if dG_total or TIR (translation initiation rate) was specified. If TIR, then convert to dG_total.\n",
        "    if TIR_target is not None:\n",
        "        dG_target = RBS_Calculator.RT_eff * (RBS_Calculator.logK - math.log(float(TIR_target)))\n",
        "\n",
        "    if verbose: print \"dG_target = \", dG_target\n",
        "\n",
        "    #Parameters\n",
        "    max_init_energy = 10.0 #kcal/mol\n",
        "    tol = 0.25 #kcal/mol\n",
        "    annealing_accept_ratios = [0.01, 0.20] #first is min, second is max\n",
        "    annealing_min_moves = 50\n",
        "    RT_init = 0.6 #roughly 300K\n",
        "\n",
        "    weighted_moves = [('insert',0.10),('delete',0.10),('replace',0.80)]\n",
        "\n",
        "    #Define the energy/cost function based on the dG_target and the other, optional targets\n",
        "    calc_energy = lambda (dG_total): abs(dG_total - dG_target)\n",
        "\n",
        "    #If RBS_Init is given, use it. Otherwise, randomly choose one that is a decent starting point.\n",
        "    if verbose: print \"Determining Initial RBS\"\n",
        "\n",
        "    if RBS_init is None:\n",
        "        (RBS,estimator) = GetInitialRBS(pre_seq,post_seq,dG_target)\n",
        "    else:\n",
        "        RBS = RBS_init\n",
        "        estimator = Run_RBS_Calculator(pre_seq,post_seq,RBS,verbose=False)\n",
        "\n",
        "    #Initialization\n",
        "    counter = 0\n",
        "    accepts = 0\n",
        "    rejects = 0\n",
        "    RT = RT_init\n",
        "\n",
        "    dG_total = estimator.dG_total_list[0]\n",
        "    energy = calc_energy(dG_total)\n",
        "\n",
        "    if verbose: print \"Initial RBS = \", RBS, \" Energy = \", energy\n",
        "    if verbose: estimator.print_dG(estimator.infinity)\n",
        "\n",
        "    while energy > tol and counter < MaxIter:\n",
        "\n",
        "        try:\n",
        "            counter += 1\n",
        "            accepted = False\n",
        "\n",
        "            move = weighted_choice(weighted_moves)\n",
        "\n",
        "            RBS_new = ''\n",
        "            if verbose: print \"Move #\", counter, \": \", move\n",
        "            if move == 'insert':\n",
        "\n",
        "                pos = int(random.uniform(0.0,1.0) * len(RBS))\n",
        "                letter = random.choice(['A', 'T', 'G', 'C'])\n",
        "                RBS_new = RBS[0:pos] + letter + RBS[pos:len(RBS)]\n",
        "\n",
        "            if move == 'delete':\n",
        "                if (len(RBS) > 1):\n",
        "                    pos = int(random.uniform(0.0,1.0) * len(RBS))\n",
        "                    RBS_new = RBS[0:pos] + RBS[pos+1:len(RBS)]\n",
        "\n",
        "            if move == 'replace':\n",
        "                pos = int(random.uniform(0.0,1.0) * len(RBS))\n",
        "                letter = random.choice(['A', 'T', 'G', 'C'])\n",
        "                RBS_new = RBS[0:pos] + letter + RBS[pos+1:len(RBS)]\n",
        "\n",
        "            RBS_new = RemoveStartCodons(RBS_new)\n",
        "\n",
        "            if len(RBS_new) > Max_RBS_Length:\n",
        "                RBS_new = RBS_new[len(RBS_new)-Max_RBS_Length:len(RBS_new)+1]\n",
        "\n",
        "            estimator = Run_RBS_Calculator(pre_seq,post_seq,RBS_new,verbose=False)\n",
        "\n",
        "            dG_total = estimator.dG_total_list[0]\n",
        "            energy_new = calc_energy(dG_total)\n",
        "            if calc_constraints(RBS_new,estimator):\n",
        "                energy_new = infinity\n",
        "\n",
        "            if verbose: print \"New energy = \", energy_new\n",
        "\n",
        "            if energy_new < energy:\n",
        "                #Accept move immediately\n",
        "                RBS = RBS_new\n",
        "                energy = energy_new\n",
        "                accepted = True\n",
        "                if verbose: print \"Move immediately accepted\"\n",
        "            else:\n",
        "\n",
        "                ddE = (energy - energy_new)\n",
        "                Metropolis = math.exp(ddE / RT)\n",
        "                prob = random.uniform(0.0,1.0)\n",
        "\n",
        "                if Metropolis > prob:\n",
        "                    #Accept move based on conditional probability\n",
        "                    RBS = RBS_new\n",
        "                    energy = energy_new\n",
        "                    accepts += 1\n",
        "                    accepted = True\n",
        "                    if verbose: print \"Move conditionally accepted\"\n",
        "                else:\n",
        "                    #Reject move\n",
        "                    rejects += 1\n",
        "                    if verbose: print \"Move rejected\"\n",
        "\n",
        "            if accepted and verbose: estimator.print_dG(estimator.infinity)\n",
        "\n",
        "            #Simulated annealing control\n",
        "            if accepts + rejects > annealing_min_moves:\n",
        "\n",
        "                ratio = float(accepts) / float(accepts + rejects)\n",
        "\n",
        "                if ratio > annealing_accept_ratios[1]:\n",
        "                    #Too many accepts, reduce RT\n",
        "                    RT = RT / 2.0\n",
        "                    accepts = 0\n",
        "                    rejects = 0\n",
        "                    if verbose: print \"Accepting too many conditional moves, reducing temperature\"\n",
        "\n",
        "                if ratio < annealing_accept_ratios[0]:\n",
        "                    #Too many rejects, increase RT\n",
        "                    RT = RT * 2.0\n",
        "                    accepts = 0\n",
        "                    rejects = 0\n",
        "                    if verbose: print \"Rejecting too many conditional moves, increasing temperature\"\n",
        "\n",
        "        except KeyboardInterrupt:\n",
        "            if verbose: print \"Calculating Final State\"\n",
        "            estimator = Run_RBS_Calculator(pre_seq,post_seq,RBS,verbose=False)\n",
        "\n",
        "            dG_total = estimator.dG_total_list[0]\n",
        "            return (dG_total, RBS, estimator, counter)\n",
        "\n",
        "    if verbose: estimator.print_dG(estimator.infinity)\n",
        "\n",
        "    if verbose: print \"Total number of RBS Evaluations: \", num_rbs_calculations\n",
        "\n",
        "    if TIR_target is not None:\n",
        "\n",
        "        #Convert back to TIR (Translation Initiation Rate)\n",
        "        TIR_out = RBS_Calculator.K * math.exp(-dG_total / RBS_Calculator.RT_eff)\n",
        "        return (TIR_out, RBS, estimator, counter)\n",
        "    else:\n",
        "        return (dG_total, RBS, estimator, counter)\n",
        "\n",
        "def MC_Design_from_file(handle, output, dG_target, verbose = True, **kvars):\n",
        "    \"\"\"This function accepts a FASTA formatted file of Pre-sequences and CDSs and generates synthetic RBS sequences with the selected target dG_total. Uses Biopython for reading FASTA files. \"\"\"\n",
        "\n",
        "    from Bio import SeqIO\n",
        "\n",
        "    #Set Defaults\n",
        "    if not \"kinetic_score_max\" in kvars.keys(): kvars[\"kinetic_score_max\"] = 0.500\n",
        "\n",
        "    records = SeqIO.parse(handle,\"fasta\")\n",
        "\n",
        "    for record in records:\n",
        "        record_list.append(record)\n",
        "\n",
        "    First = True\n",
        "    counter=0\n",
        "\n",
        "    assert len(dG_target) == len(record_list), \"The length of dG_target should equal the number of designed RBSs in the input file\"\n",
        "\n",
        "    while counter < len(record_list):\n",
        "        pre_seq = record_list[counter].seq.tostring().upper()\n",
        "        post_seq = record_list[counter+1].seq.tostring().upper()\n",
        "        counter += 2\n",
        "\n",
        "        [dG_total, RBS, estimator] = Monte_Carlo_Design(pre_seq, post_seq, dG_target[counter/2-1], verbose,kvars)\n",
        "\n",
        "        if verbose: estimator.print_dG(estimator.infinity)\n",
        "        if verbose: print \"Final RBS = \", RBS\n",
        "\n",
        "        estimator.save_data(output, First)\n",
        "        if First:\n",
        "            First = False\n",
        "\n",
        "        index = estimator.mRNA_rRNA_corrected_structure_list[0][\"MinStructureID\"]\n",
        "        estimator.mRNA_rRNA_corrected_structure_list[0].export_PDF(index, \"Monte Carlo Result\", \"MC_\" + str(counter/2-1) + \"_rRNA\" + \".pdf\")\n",
        "\n",
        "        estimator.mRNA_structure_list[0].export_PDF(0, \"Monte Carlo Result\", \"MC_\" + str(counter/2-1) + \"_mRNA\" + \".pdf\")\n",
        "\n",
        "#-------------------------------------------------------------------------------\n",
        "if __name__ == \"__main__\":\n",
        "\n",
        "\n",
        "    pre_RBS = \"TTCTAGA\"\n",
        "    post_RBS = \"ATGCAGCACGTGTGCAGCACTACAGCGTGTGACGACTACAGCATTCACGACAGTCACATGCAGTTGACAC\"\n",
        "    dG_target = -10.0\n",
        "\n",
        "    (dG_total, RBS, estimator, iterations) = Monte_Carlo_Design(pre_RBS, post_RBS, RBS_init = None, dG_target = dG_target, MaxIter = 10000, verbose = True)\n",
        "\n",
        "    print \"Finished\"\n",
        "    print \"dG_total = \", dG_total\n",
        "    print \"RBS = \", RBS\n",
        "    print \"# iterations = \", iterations\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "49993GCay3t3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title code of Run_RBS_Design\n",
        "#!/usr/bin/env python\n",
        "\n",
        "#This file is part of the Ribosome Binding Site Calculator.\n",
        "\n",
        "#The Ribosome Binding Site Calculator is free software: you can redistribute it and/or modify\n",
        "#it under the terms of the GNU General Public License as published by\n",
        "#the Free Software Foundation, either version 3 of the License, or\n",
        "#(at your option) any later version.\n",
        "\n",
        "from RBS_MC_Design import Monte_Carlo_Design\n",
        "import sys, time\n",
        "\n",
        "def all(x):\n",
        "    \"\"\"The all() function from Python 2.5, for backwards compatibility with previous versions of Python.\"\"\"\n",
        "\n",
        "    output = True\n",
        "    for bool in x:\n",
        "        if not bool:\n",
        "            output = False\n",
        "            break\n",
        "    return output\n",
        "\n",
        "def run():\n",
        "\n",
        "    #Read command line arguments\n",
        "    input_args = []\n",
        "    for arg in sys.argv:\n",
        "        input_args.append(arg)\n",
        "\n",
        "    if len(input_args) == 4:\n",
        "        pre_seq = input_args[1]\n",
        "        post_seq = input_args[2]\n",
        "        TIR = float(input_args[3])\n",
        "    else:\n",
        "        output = \"Usage: \" + input_args[0] + \" [Pre-Sequence] [Post-Sequence] [Translation Initiation Rate]\" + \"\\n\"\n",
        "        return (False,output)\n",
        "\n",
        "    #Check Sequence and TIR Inputs\n",
        "    valid_chars = ('A','C','T','G','U','a','c','t','g','u')\n",
        "    valid_start_codons = ('ATG','AUG','GTG','GUG')\n",
        "\n",
        "    if TIR <= 0: return (False,\"invalid TIR\")\n",
        "    if len(pre_seq) <= 0 or not all([s in valid_chars for s in pre_seq[:]]): return (False,\"invalid pre_seq\")\n",
        "    if len(post_seq) <= 0 or not all([s in valid_chars for s in post_seq[:]]) or post_seq[0:3].upper() not in valid_start_codons: return (False,\"invalid post_seq\")\n",
        "\n",
        "    total_iterations=0\n",
        "    MaxIter = 10000\n",
        "    (TIR_out, RBS_out, estimator, iterations) = Monte_Carlo_Design(pre_seq,post_seq, None, TIR, MaxIter, verbose = False)\n",
        "    total_iterations+=iterations\n",
        "\n",
        "    #If RBS is not designed to specs within 10000 iterations, then start from a different initial condition\n",
        "    while iterations == MaxIter:\n",
        "        (TIR_out, RBS_out, estimator, iterations) = Monte_Carlo_Design(pre_seq,post_seq, None, TIR, MaxIter, verbose = False)\n",
        "        total_iterations+=iterations\n",
        "\n",
        "    output_string = \"Program Executed\" + \"\\n\" + RBS_out + \"\\n\" + str(TIR_out) + \"\\n\" + str(total_iterations) + \"\\n\"\n",
        "    return (True,output_string)\n",
        "\n",
        "def ReadOutput(output):\n",
        "    import stringIO\n",
        "\n",
        "    handle = stringIO.stringIO(output)\n",
        "\n",
        "    for line in handle.readlines():\n",
        "        if line == \"Program Executed\":\n",
        "            RBS_out = handle.readlines()\n",
        "            TIR_out = float(handle.readlines())\n",
        "            total_iterations = int(handle.readlines())\n",
        "            break\n",
        "        elif line == \"Program NOT Executed\":\n",
        "            RBS_out = \"\"\n",
        "            TIR_out = -1.0\n",
        "            total_iterations = 0\n",
        "            break\n",
        "\n",
        "    return (RBS_out, TIR_out, total_iterations)\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "\n",
        "    #Runs RBS Design optimization\n",
        "    #Prints to standard output -- captured by globusrun interactive and transferred to the submitting computer.\n",
        "    start = time.clock()\n",
        "\n",
        "    (success,output) = run()\n",
        "\n",
        "    end = time.clock()\n",
        "    time_elapsed = end - start\n",
        "\n",
        "    if success:\n",
        "        print output\n",
        "    else:\n",
        "        print \"Error\"\n",
        "        print output\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "AvO8V7BMz1z_"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}